from pandas import read_csv

ns = read_csv('/mnt/HDD3/lrma/script/ns_subg.csv', index_col=0)
ns = ns.loc[ns['identity_filter']]

def get_subg(S):
    return [f'/home/mathieu/paradoxus_nanopore/MA_parents/assemblies/{ns.loc[s,"subg"]}.chromosomes.rdna.fasta' for s in S]

rule all:
    input:
        expand('/mnt/HDD3/lrma/svim/{s}/variants.vcf', s=ns['s_subg'].values)

rule svim:
    input:
        reads = '/mnt/HDD3/lrma/sort/{s}.reads.fastq',
        ref = get_subg
    params:
        outdir = '/mnt/HDD3/lrma/svim/{s}/'
    output:
        '/mnt/HDD3/lrma/svim/{s}/variants.vcf'
    shell:
        'if [ ! -d {params.outdir} ]; then mkdir {params.outdir}; fi; '
        'svim reads {params.outdir} {input.reads} {input.ref} --aligner minimap2 --nanopore'
