{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b54122de-db6c-413c-960c-c25a273040a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('/home/mathieu/software/pysal')\n",
    "import pysalvador as sal\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.patches import FancyArrow\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "from matplotlib.ticker import LogLocator\n",
    "from PIL import Image\n",
    "import seaborn as sns\n",
    "sns.set(style='ticks', font='DejaVu Sans')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc7e689f-31bb-4551-a404-9526a58644b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#path for figures\n",
    "fig_path = '/home/mathieu/mhenault_landrylab/Publications/lrma/resubmission1_elife/fig/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21dd0344-80ec-430e-8e8a-c4ee391874fd",
   "metadata": {},
   "source": [
    "## Import haploids metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45c461d8-dbd5-4048-ba2b-ee1373d938a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "haploids = pd.read_excel('/home/mathieu/mhenault_landrylab/Experiments/transposition_assay_2022/strains_crosses_3.ods',\n",
    "                         sheet_name='haploids', header=None, index_col=0)\n",
    "haploids.columns = ['strain','genotype','description','','mat','media','box','pos','date','person','ref']\n",
    "haploids['background'] = haploids['strain'].apply(lambda x: x.split(' ')[0])\n",
    "cross_plate = pd.read_csv('/home/mathieu/mhenault_landrylab/Experiments/transposition_assay_2022/crosses_plate.csv', index_col=0)\n",
    "cn_ty1 = pd.read_csv('/home/mathieu/mhenault_landrylab/Experiments/transposition_assay_2021/cn_ty1.csv')\n",
    "cn_ty1.loc[1, 'strain'] = 'UWOPS-79-140'\n",
    "cn_ty1.index = cn_ty1['strain'].values\n",
    "\n",
    "backgrounds = ['YPS744','UWOPS-79-140','UWOPS-91-202','GB23','LL17-69','PPC6','VC2','D11','GB20','LL17-02','PPC3','SQ107','IO33','PPC20']\n",
    "background_alias = dict(zip(backgrounds, ['YPS744','UWOPS-79-140','UWOPS-91-202','MSH-604','LL2012_014','LL2012_021','LL2012_028','LL2011_012','MSH-587-1',\\\n",
    "                             'LL2012_011','LL2012_019','LL2011_009','LL2011_005','LL2012_016']))\n",
    "bgd_color = dict(zip(backgrounds, [\"#00CD00\",\"#EE0000\",\"#EE0000\",\"#EE0000\",\"#EE0000\",\"#EE0000\",\"#EE0000\",\"#00008B\",\"#00008B\",\"#00008B\",\"#00008B\",\"#00008B\",\"#912CEE\",\"#912CEE\"]))\n",
    "bgd_group = dict(zip(backgrounds, [\"SpA\",\"SpB\",\"SpB\",\"SpB\",\"SpB\",\"SpB\",\"SpB\",\"SpC\",\"SpC\",\"SpC\",\"SpC\",\"SpC\",\"SpC*\",\"SpC*\"]))\n",
    "group_alias = {'SpA':r'$SpA$','SpB':r'$SpB$','SpC':r'$SpC$','SpC*':r'$SpC^*$'}\n",
    "\n",
    "\n",
    "hybrids = [frozenset(['YPS744']),\n",
    "           frozenset(['UWOPS-79-140']),\n",
    "           frozenset(['D11']),\n",
    "           frozenset(['YPS744', 'UWOPS-79-140']),\n",
    "           frozenset(['YPS744', 'D11']),\n",
    "           frozenset(['D11', 'UWOPS-79-140'])]\n",
    "           \n",
    "hybrid_colors = dict(zip(hybrids, [\"#00CD00\",\n",
    "                                   \"#EE0000\",\n",
    "                                   \"#00008B\", \n",
    "                                   np.array([168, 145, 0])/255, \n",
    "                                   np.array([0, 145, 98])/255,\n",
    "                                   np.array([168, 0, 98])/255]\n",
    "                                   ))\n",
    "\n",
    "for bgd, df in haploids.groupby('background'):\n",
    "    if bgd in cn_ty1.index:\n",
    "        haploids.loc[df.index, 'cn'] = cn_ty1.loc[bgd,'median_corr']\n",
    "haploids['group'] = haploids['background'].apply(lambda x: bgd_group[x])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23b5fa52-502a-4481-8c71-b9f1f9023b5e",
   "metadata": {},
   "source": [
    "## Import hybrids metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98c606e9-1081-43c7-9371-2914c9f02a0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "cross_rho0 = pd.read_excel('/home/mathieu/mhenault_landrylab/Experiments/transposition_assay_2022/strains_crosses_3.ods',\n",
    "                         sheet_name='cross_rho0', header=0, index_col=0).set_index('hybrid_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b1250e8-326d-4ab0-a602-2679a53cbf8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# export plate ploidy 1\n",
    "\n",
    "ploidy1 = cross_rho0.loc[cross_rho0['plate_ploidy']==1].copy().fillna('')\n",
    "ploidy1['value'] = ploidy1.apply(lambda x: f'{x[\"src_well\"]}.{x[\"colony\"]}', axis=1)\n",
    "ploidy1.pivot_table(index='row', columns='col', values='value', aggfunc=lambda x: x)\\\n",
    ".to_csv('/home/mathieu/mhenault_landrylab/Experiments/transposition_assay_2022/ploidy1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87f614ab-6566-4503-b538-3ea5488f9d43",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rates_table(results, meta):\n",
    "    \n",
    "    rates = []\n",
    "\n",
    "    wells = results.index[5:]\n",
    "\n",
    "    for i in results.columns:\n",
    "        s, plasmid, date, count, N = results.loc[['strain_id', 'plasmid', 'date', 'count', 'N'], i]\n",
    "        counts = results.loc[wells, i]\n",
    "        counts = counts[~np.isnan(counts.astype(np.float32))].astype(np.int32)\n",
    "\n",
    "        if counts.shape[0] > 3:\n",
    "\n",
    "            if ~np.all(counts==0):\n",
    "                upper_bound = False\n",
    "            else:\n",
    "                counts[0] = 1\n",
    "                upper_bound = True\n",
    "\n",
    "            n = sal.newtonLD(counts)\n",
    "            ci = sal.confintLD(counts, show_iter=False)\n",
    "\n",
    "            rates.append([s, plasmid, date, count, N, n, ci[0], ci[1], upper_bound])\n",
    "\n",
    "    rates = pd.DataFrame(rates, columns=['strain_id', 'plasmid', 'date', 'count', 'N', 'n', 'n_0.025', 'n_0.975', 'upper_bound'])\n",
    "    rates['rate'] = rates['n']/rates['N']\n",
    "    rates['rate_0.025'] = rates['n_0.025']/rates['N']\n",
    "    rates['rate_0.975'] = rates['n_0.975']/rates['N']\n",
    "    rates.set_index('strain_id', inplace=True)\n",
    "\n",
    "    # append metadata\n",
    "    for c in meta.columns:\n",
    "        rates[c] = meta.loc[rates.index, c].values\n",
    "        \n",
    "    return rates.reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fa98ef3-8b59-484d-8d6c-930ee722ca12",
   "metadata": {},
   "source": [
    "# Haploids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65116819-29b1-463e-91ef-d61024cfd9fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_haploids = pd.read_excel('/home/mathieu/mhenault_landrylab/Experiments/transposition_assay_2022/data_2023/results.xlsx',\n",
    "                                 sheet_name='haploids', header=None, index_col=0)\n",
    "rates_haploids = rates_table(results_haploids, haploids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb13f666-63d1-446e-8630-22560768b5a4",
   "metadata": {},
   "source": [
    "## Fig 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00482987-d465-4d8e-b210-3e430348585d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=[6,5], gridspec_kw=dict(right=0.95, top=0.75, bottom=0.12, left=0.2))\n",
    "\n",
    "ylim = [1e-10, 3e-6]\n",
    "\n",
    "for i in rates_haploids.index:\n",
    "    bgd, plasmid, group, rate, rate_0025, rate_0975, upper, cn = \\\n",
    "    rates_haploids.loc[i, ['background', 'plasmid', 'group', 'rate', 'rate_0.025', 'rate_0.975', 'upper_bound', 'cn']]\n",
    "    \n",
    "    mec = bgd_color[bgd]\n",
    "    \n",
    "    fc = 'magenta'\n",
    "    if upper:\n",
    "        fc = 'white'\n",
    "        \n",
    "    else:\n",
    "        fc = mec\n",
    "        \n",
    "    m = 'x'\n",
    "    if plasmid == 'SpB':\n",
    "        m = 'o'\n",
    "        x = cn*0.98\n",
    "    elif plasmid == 'SpC':\n",
    "        m = 's'\n",
    "        x = cn*1.02\n",
    "        \n",
    "    if not upper:\n",
    "        ax.plot([x, x], [rate_0025, rate_0975], color='k', zorder=1)\n",
    "    ax.scatter(x, rate, marker=m, s=36, color=fc, edgecolors=mec, zorder=2)\n",
    "    \n",
    "bgd_rate_cn = rates_haploids.groupby(['background', 'cn'])['rate'].min().reset_index().sort_values(by='cn').set_index('background')\n",
    "bgd_rate_cn['cn_rank'] = np.logspace(-5, -7, 8)\n",
    "\n",
    "#add background labels\n",
    "\"\"\"for bgd in bgd_rate_cn.index:\n",
    "    cn, rate, rank = bgd_rate_cn.loc[bgd]\n",
    "    \n",
    "    fa = FancyArrow(cn, ylim[0], 0, ylim[1]-ylim[0], width=0.15*cn, head_length=0, head_width=0,\n",
    "                    fc='0.5', lw=0, alpha=0.3, zorder=0)\n",
    "    ax.add_patch(fa)\n",
    "    if rate > 7e-8:\n",
    "        y = 3e-9\n",
    "    else:\n",
    "        y = 3e-7\n",
    "    ax.text(cn, y, background_alias[bgd], rotation=90, va='center', ha='center', size=8)\"\"\"\n",
    "    \n",
    "ax.set_ylim(ylim)\n",
    "ax.set_yscale('log')\n",
    "ax.set_ylabel('Transposition rate\\n(cell$^{-1}$ gen$^{-1}$)')\n",
    "ax.set_xlabel('Chromosomal Ty1 CN')\n",
    "ax.set_xscale('log')\n",
    "ax.set_xticks([1,2,5,10,20, 30])\n",
    "ax.set_xticklabels([1,2,5,10,20, 30])\n",
    "\n",
    "legend_upper = 1.37\n",
    "\n",
    "legend_elms0 = [Line2D([0], [0], color='w', marker='o', mfc=fc, ms=8, label=group_alias[l]) for (l, fc) in zip(['SpA','SpB','SpC'], ['#00CD00', '#EE0000', '#00008B'])]\n",
    "legend0 = plt.legend(handles=legend_elms0, title='Haploid background', bbox_to_anchor=[-0.25, legend_upper], loc='upper left')\n",
    "\n",
    "legend_elms1 = [Line2D([0], [0], color='w', marker=m, mfc='k', mec='k', ms=8, label=group_alias[l]) for (l, m) in zip(['SpB','SpC'], ['o', 's'])]\n",
    "legend1 = plt.legend(handles=legend_elms1, title='Ty1 plasmid', bbox_to_anchor=[0.3, legend_upper], loc='upper left')\n",
    "\n",
    "legend_elms2 = [Line2D([0], [0], color='w', marker='o', mfc=fc, mec='k', ms=8, label=l) for (l, fc) in zip(['No','Yes'], ['k', 'w'])]\n",
    "legend2 = plt.legend(handles=legend_elms2, title='Upper bound', bbox_to_anchor=[0.7, legend_upper], loc='upper left')\n",
    "\n",
    "for legend in [legend0, legend1, legend2]:\n",
    "    ax.add_artist(legend)\n",
    "\n",
    "\n",
    "sns.despine()\n",
    "\n",
    "#for ext in ['svg', 'png']:\n",
    "#    plt.savefig(f'{fig_path}Fig5.{ext}', dpi=300)\n",
    "    \n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "863efc1d-44f1-41b8-b3e6-3569f8202b48",
   "metadata": {},
   "outputs": [],
   "source": [
    "rates_old = pd.read_csv('/home/mathieu/mhenault_landrylab/Experiments/transposition_assay_2021/res/rates_061221.csv', index_col=0)\n",
    "rates_old.set_index(['cross','plasmid'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dd56c42-d36a-4341-abee-87a8438b5fb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (p1, p2, plasmid), df in rates.groupby(['parent1', 'parent2', 'plasmid']):\n",
    "    key1 = (f'{p1}.{p2}', plasmid[-1])\n",
    "    key2 = (f'{p2}.{p1}', plasmid[-1])\n",
    "    \n",
    "    if key1 in rates_old.index:\n",
    "        ro = rates_old.loc[key1, 'rate']\n",
    "        rates.loc[df.index, 'rate_old'] = ro\n",
    "        \n",
    "    elif key2 in rates_old.index:\n",
    "        ro = rates_old.loc[key2, 'rate']\n",
    "        rates.loc[df.index, 'rate_old'] = ro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04999f9d-d689-4c98-93fe-d90d5ee78ae4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(x='rate', y='rate_old', data=rates)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71810e8d-27fd-4455-a0ed-2d73c8583b55",
   "metadata": {},
   "source": [
    "# Hybrids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b278697c-a3bf-4ba2-a886-fbe673e4f0c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_hybrids = pd.read_excel('/home/mathieu/mhenault_landrylab/Experiments/transposition_assay_2022/data_2023/results.xlsx',\n",
    "                                 sheet_name='hybrids', header=None, index_col=0)\n",
    "rates_hybrids = rates_table(results_hybrids, cross_rho0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f1605e3-07bc-4575-bbc7-9f2f290b92f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dictionary of crosses labels for plotting\n",
    "mat_alias = {'a': r'$a$', 'alpha':r'$\\alpha$'}\n",
    "rho_alias = {'rho+': r'$\\rho^+$', 'rho0':r'$\\rho^0$'}\n",
    "\n",
    "hybrid_labels_rho = {}\n",
    "\n",
    "for idx, df in rates_hybrids.groupby(['parent1_background', 'parent1_mat', 'parent1_rho', 'parent2_background', 'parent2_mat', 'parent2_rho']):\n",
    "    (bgd1, mat1, rho1, bgd2, mat2, rho2) = idx\n",
    "\n",
    "    times = r'$\\times$'\n",
    "    label = f\"{background_alias[bgd2]} $MAT${mat_alias[mat2]} {rho_alias[rho2]}{times}\\n{background_alias[bgd1]} $MAT${mat_alias[mat1]} {rho_alias[rho1]}\"\n",
    "    \n",
    "    hybrid_labels_rho[idx] = label\n",
    "    \n",
    "hybrid_labels_mat ={}\n",
    "for idx, df in rates_hybrids.groupby(['parent1_background', 'parent1_mat', 'parent2_background', 'parent2_mat']):\n",
    "    (bgd1, mat1, bgd2, mat2) = idx\n",
    "\n",
    "    times = r'$\\times$'\n",
    "    label = f\"{background_alias[bgd2]} $MAT${mat_alias[mat2]} ({group_alias[bgd_group[bgd2]]}){times}\\n{background_alias[bgd1]} $MAT${mat_alias[mat1]} ({group_alias[bgd_group[bgd1]]})\"\n",
    "    \n",
    "    hybrid_labels_mat[idx] = label\n",
    "    \n",
    "hybrid_labels ={}\n",
    "for idx, df in rates_hybrids.groupby(['parent1_background', 'parent1_mat', 'parent2_background', 'parent2_mat']):\n",
    "    (bgd1, mat1, bgd2, mat2) = idx\n",
    "\n",
    "    times = r'$\\times$'\n",
    "    label = f\"{background_alias[bgd2]} ({group_alias[bgd_group[bgd2]]}){times}\\n{background_alias[bgd1]} ({group_alias[bgd_group[bgd1]]})\"\n",
    "    \n",
    "    hybrid_labels[idx] = label"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb1960c6-5d36-4ba3-a6e2-3ff05490050b",
   "metadata": {},
   "source": [
    "## Fig S13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9026c5c3-fe3a-4ac2-845c-54a2d1ea6fc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_cross_parents(df, s, X, ax, dx=0.2, color='k'):\n",
    "    \n",
    "    global hybrid_labels_rho\n",
    "    \n",
    "    for plasmid in ['SpB', 'SpC']:\n",
    "    \n",
    "        rate, rate_0025, rate_0975, upper = df.loc[plasmid, ['rate', 'rate_0.025', 'rate_0.975', 'upper_bound']]\n",
    "        \n",
    "        fc = 'magenta'\n",
    "        if upper:\n",
    "            fc = 'white'\n",
    "\n",
    "        else:\n",
    "            fc = color\n",
    "        \n",
    "        m = 'x'\n",
    "        x = np.nan\n",
    "        if plasmid == 'SpB':\n",
    "            m = 'o'\n",
    "            x =  X-dx\n",
    "            \n",
    "        elif plasmid == 'SpC':\n",
    "            m = 's'\n",
    "            x = X+dx\n",
    "\n",
    "        if not upper:\n",
    "            ax.plot([x, x], [rate_0025, rate_0975], color='k', zorder=1)\n",
    "        ax.scatter(x, rate, marker=m, s=48, color=fc, edgecolors=color, zorder=2)\n",
    "        \n",
    "    idx = df.value_counts(['parent1_background', 'parent1_mat', 'parent1_rho', 'parent2_background', 'parent2_mat', 'parent2_rho']).index[0]\n",
    "    \n",
    "    return hybrid_labels_rho[idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2b60437-ca70-4ff3-ae70-7afa3b8561ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(figsize=[10,7], ncols=3, \n",
    "                        gridspec_kw=dict(left=0.1, right=0.97, top=0.7, bottom=0.3, wspace=0.35))\n",
    "    \n",
    "hyb_included = {hyb:np.nan for hyb in hybrids}\n",
    "\n",
    "for (bgd1, bgd2), ax_idx in zip(itertools.combinations(['YPS744', 'UWOPS-79-140', 'D11'], 2), range(3)):\n",
    "\n",
    "    #bgd1, bgd2 = sorted([bgd1, bgd2])\n",
    "    \n",
    "    ax = axes[ax_idx]\n",
    "    x_labels = {}\n",
    "    \n",
    "    for i, (p1, p2) in zip(range(3),\n",
    "                              [(bgd1, bgd1), (bgd1, bgd2), (bgd2, bgd2)]):\n",
    "        \n",
    "        df = rates_hybrids.loc[(rates_hybrids['parent1_background']==p2) & (rates_hybrids['parent2_background']==p1)]\n",
    "        hyb = frozenset([p1, p2])\n",
    "        \n",
    "        idx = tuple(df.iloc[0][['parent1_background', 'parent1_mat', 'parent2_background', 'parent2_mat']])    \n",
    "        hyb_included[hyb] = idx\n",
    "        \n",
    "        # split in hybrid genotypes\n",
    "        \n",
    "        #extract how many genotypes exist for that hybrid\n",
    "        uhyb = df.sort_values(by='parent1_rho')['strain_id'].unique()\n",
    "        nuhyb = uhyb.shape[0]\n",
    "        #determine plotting order of the different crosses\n",
    "        if i == 1:\n",
    "            s_order = dict(zip(uhyb, np.arange(nuhyb)))\n",
    "        elif i == 0:\n",
    "            s_order = dict(zip(uhyb, (np.arange(nuhyb)+1)*(-1)))\n",
    "        elif i == 2:\n",
    "            s_order = dict(zip(uhyb, np.arange(nuhyb)+2))       \n",
    "        \n",
    "        # get the color of the cross\n",
    "        color = hybrid_colors[frozenset([p1, p2])]\n",
    "        \n",
    "        for s, df1 in df.groupby('strain_id'):\n",
    "            df1 = df1.set_index('plasmid')\n",
    "            \n",
    "            x = s_order[s]\n",
    "            l = plot_cross_parents(df1, s, x, ax, dx=0.1, color=color)\n",
    "            x_labels[x] = l\n",
    "            \n",
    "    x_ticks = sorted(x_labels.keys())\n",
    "    ax.set_xticks(x_ticks)\n",
    "    ax.set_xticklabels([x_labels[x] for x in x_ticks], rotation=90, ha='center', va='top', size=10)\n",
    "    \n",
    "    ax.set_yscale('log')\n",
    "    ax.set_ylim(5e-11, 2e-6)\n",
    "    if ax_idx == 0:\n",
    "        ax.set_ylabel('Transposition rate (cell$^{-1}$ gen$^{-1}$)')\n",
    "    \n",
    "    sns.despine(ax=ax)\n",
    "    \n",
    "# draw legends\n",
    "ax = fig.add_axes([0,1,1,0])\n",
    "ax.axis('off')\n",
    "\n",
    "legend_elms0 = [Line2D([0], [0], color='w', marker='o', mfc=hybrid_colors[k], ms=10, \n",
    "                        label=hybrid_labels[idx]) for k,idx in hyb_included.items()]\n",
    "legend0 = plt.legend(handles=legend_elms0, title='Diploid background', bbox_to_anchor=[0, 1], loc='upper left', ncol=2)\n",
    "\n",
    "legend_elms1 = [Line2D([0], [0], color='w', marker=m, mfc='k', mec='k', ms=10, label=group_alias[l]) for (l, m) in zip(['SpB','SpC'], ['o', 's'])]\n",
    "legend1 = plt.legend(handles=legend_elms1, title='Ty1 plasmid', bbox_to_anchor=[0.6, 1], loc='upper left')\n",
    "\n",
    "legend_elms2 = [Line2D([0], [0], color='w', marker='o', mfc=fc, mec='k', ms=10, label=l) for (l, fc) in zip(['No','Yes'], ['k', 'w'])]\n",
    "legend2 = plt.legend(handles=legend_elms2, title='Upper bound', bbox_to_anchor=[0.85, 1], loc='upper left')\n",
    "\n",
    "for legend in [legend0, legend1, legend2]:\n",
    "    ax.add_artist(legend)\n",
    "    \n",
    "#for ext in ['svg','png']:\n",
    "#    plt.savefig(f'{fig_path}Supp_FigS12.{ext}', dpi=300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30ad28f6-5775-4b14-8eb1-e0af6b29f681",
   "metadata": {},
   "source": [
    "# Correlations between rates for $\\rho^0$ reciprocal crosses\n",
    "## Fig 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dba2e42-4617-4ae9-adbc-f7c24a0a6302",
   "metadata": {},
   "outputs": [],
   "source": [
    "#fig, ax = plt.subplots(figsize=[8,4.5], gridspec_kw=dict(left=0.15, bottom=0.2, right=0.55, top=0.93))\n",
    "fig = plt.figure(figsize=[8, 4.5])\n",
    "gs = plt.GridSpec(ncols=1, nrows=1, left=0.15, bottom=0.2, right=0.55, top=0.93)\n",
    "ax = fig.add_subplot(gs[0])\n",
    "\n",
    "#hyb_included = {}\n",
    "hyb_included = {hyb:np.nan for hyb in hybrids}\n",
    "xlim = np.array([7e-11, 2e-7])\n",
    "\n",
    "for hyb in hybrids:\n",
    "    Hyb = sorted(list(hyb))\n",
    "    if len(Hyb) == 1:\n",
    "        Hyb = [Hyb[0], Hyb[0]]\n",
    "    \n",
    "    df = rates_hybrids.loc[(rates_hybrids['parent1_background']==Hyb[0]) & (rates_hybrids['parent2_background']==Hyb[1])]\n",
    "    \n",
    "    if df['parent1_rho'].unique().shape[0] > 1:\n",
    "        \n",
    "        \n",
    "        for plasmid, df1 in df.sort_values(by='parent1_rho').groupby('plasmid'):\n",
    "            \n",
    "            rate1, rate1_0025, rate1_0975, rate1_upper = df1.iloc[0][['rate', 'rate_0.025', 'rate_0.975', 'upper_bound']]\n",
    "            rate2, rate2_0025, rate2_0975, rate2_upper = df1.iloc[1][['rate', 'rate_0.025', 'rate_0.975', 'upper_bound']]\n",
    "            \n",
    "            idx = tuple(df1.iloc[0][['parent1_background', 'parent1_mat', 'parent2_background', 'parent2_mat']])\n",
    "            \n",
    "            hyb_included[hyb] = idx\n",
    "            \n",
    "            if plasmid == 'SpB':\n",
    "                m = 'o'\n",
    "\n",
    "            elif plasmid == 'SpC':\n",
    "                m = 's'\n",
    "                \n",
    "            \n",
    "            if any([rate1_upper, rate2_upper]):\n",
    "                mfc = 'w'\n",
    "            else:\n",
    "                mfc = hybrid_colors[hyb]\n",
    "            \n",
    "            ax.scatter(rate2, rate1, marker=m, zorder=2, color=mfc, edgecolors=hybrid_colors[hyb])\n",
    "            \n",
    "            if not rate2_upper:\n",
    "                ax.plot([rate2_0025, rate2_0975], [rate1, rate1], color='k', zorder=1)\n",
    "            if not rate1_upper:\n",
    "                ax.plot([rate2, rate2], [rate1_0025, rate1_0975], color='k', zorder=1)\n",
    "\n",
    "ax.set_xscale('log')\n",
    "ax.set_xlim(xlim)\n",
    "ax.set_xlabel('Transposition rate (cell$^{-1}$ gen$^{-1}$)\\n$MAT$'+mat_alias[\"a\"]+' parent mtDNA')\n",
    "\n",
    "ax.set_yscale('log')\n",
    "ax.set_ylim(xlim)\n",
    "ax.set_ylabel('Transposition rate (cell$^{-1}$ gen$^{-1}$)\\n$MAT$'+mat_alias[\"alpha\"]+' parent mtDNA')\n",
    "\n",
    "ax.plot(xlim, xlim, zorder=0, color='k', lw=0.5)\n",
    "\n",
    "# Legends\n",
    "legend_left = 1.1\n",
    "\n",
    "legend_elms0 = [Line2D([0], [0], color='w', marker='o', mfc=hybrid_colors[k], ms=10, \n",
    "                       label=hybrid_labels_mat[idx]) for k,idx in hyb_included.items() if type(idx)==tuple]\n",
    "legend0 = plt.legend(handles=legend_elms0, title='Diploid background', bbox_to_anchor=[legend_left, 1.1], loc='upper left', ncol=1)\n",
    "\n",
    "legend_elms1 = [Line2D([0], [0], color='w', marker=m, mfc='k', mec='k', ms=8, label=l) for (l, m) in zip(['$SpB$','$SpC$'], ['o', 's'])]\n",
    "legend1 = plt.legend(handles=legend_elms1, title='Ty1 plasmid', bbox_to_anchor=[legend_left, 0.43], loc='upper left')\n",
    "\n",
    "legend_elms2 = [Line2D([0], [0], color='w', marker='o', mfc=fc, mec='k', ms=8, label=l) for (l, fc) in zip(['No','Yes'], ['k', 'w'])]\n",
    "legend2 = plt.legend(handles=legend_elms2, title='Upper bound', bbox_to_anchor=[legend_left, 0.16], loc='upper left')\n",
    "\n",
    "for legend in [legend0, legend1, legend2]:\n",
    "    ax.add_artist(legend)\n",
    "\n",
    "sns.despine()\n",
    "\n",
    "# add new axis for labels\n",
    "ax1 = fig.add_subplot(gs[0])\n",
    "ax1.axis('off')\n",
    "\n",
    "dxy = 0.15\n",
    "for hyb in hybrids:\n",
    "    if len(hyb) == 2:\n",
    "        Hyb = sorted(list(hyb))\n",
    "        fa = FancyArrow(dxy, dxy, 0.1, -0.1, width=0.08, lw=0, length_includes_head=True, zorder=-1,\n",
    "                       head_length=0.02, head_width=0.08, fc=hybrid_colors[hyb], alpha=0.7)\n",
    "        ax1.add_patch(fa)\n",
    "        ax1.text(dxy+0.05, dxy-0.05, group_alias[bgd_group[Hyb[1]]], rotation=315, ha='center', va='center', size=10, c='w')\n",
    "        \n",
    "        fa = FancyArrow(dxy, dxy, -0.1, +0.1, width=0.08, lw=0, length_includes_head=True, zorder=-1,\n",
    "                       head_length=0.02, head_width=0.08, fc=hybrid_colors[hyb], alpha=0.7)\n",
    "        ax1.add_patch(fa)\n",
    "        ax1.text(dxy-0.05, dxy+0.05, group_alias[bgd_group[Hyb[0]]], rotation=315, ha='center', va='center', size=10, c='w')\n",
    "        \n",
    "        dxy += 0.065\n",
    "\n",
    "#for ext in ['svg', 'png']:\n",
    "#    plt.savefig(f'{fig_path}Fig7.{ext}', dpi=300)\n",
    "\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc4b90fc-db9c-4434-9a85-2c4fea59e360",
   "metadata": {},
   "source": [
    "# Correlations between rates for Ty1 plasmids\n",
    "## Fig 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7be3928d-58da-46e4-980c-015d99b2d7b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=2, figsize=[7.5,8], \n",
    "                         gridspec_kw=dict(left=0.16, bottom=0.1, right=0.6, top=0.94, hspace=0.4))\n",
    "\n",
    "## Haploids\n",
    "\n",
    "ax = axes[0]\n",
    "xlim = np.array([5e-11, 3e-6])\n",
    "\n",
    "\n",
    "for bgd, df in rates_haploids.groupby('background'):\n",
    "\n",
    "    if df['plasmid'].unique().shape[0] > 1:\n",
    "        \n",
    "        df = df.set_index('plasmid')\n",
    "        \n",
    "        b_rate, b_0025, b_0975, b_upper = df.loc['SpB', ['rate', 'rate_0.025', 'rate_0.975', 'upper_bound']]\n",
    "        c_rate, c_0025, c_0975, c_upper = df.loc['SpC', ['rate', 'rate_0.025', 'rate_0.975', 'upper_bound']]\n",
    "        \n",
    "        if any([b_upper, c_upper]):\n",
    "            mfc = 'w'\n",
    "        else:\n",
    "            mfc = bgd_color[bgd]\n",
    "        \n",
    "        ax.scatter(b_rate, c_rate, marker='o', color=mfc, edgecolors=bgd_color[bgd], zorder=2)\n",
    "        \n",
    "        if not b_upper:\n",
    "            ax.plot([b_0025, b_0975], [c_rate, c_rate], color='k', zorder=1)\n",
    "        if not c_upper:\n",
    "            ax.plot([b_rate, b_rate], [c_0025, c_0975], color='k', zorder=1)\n",
    "    \n",
    "ax.set_xscale('log')\n",
    "ax.set_xlim(xlim)\n",
    "ax.set_xlabel('Transposition rate (cell$^{-1}$ gen$^{-1}$)\\n'+group_alias['SpB']+' Ty1 variant')\n",
    "\n",
    "ax.set_yscale('log')\n",
    "ax.set_ylim(xlim)\n",
    "ax.set_ylabel('Transposition rate (cell$^{-1}$ gen$^{-1}$)\\n'+group_alias['SpC']+' Ty1 variant')\n",
    "\n",
    "ax.plot(xlim, xlim, zorder=0, color='k', lw=0.5)\n",
    "    \n",
    "## Hybrids\n",
    "\n",
    "hyb_included = {hyb:np.nan for hyb in hybrids}\n",
    "\n",
    "ax = axes[1]\n",
    "for (bgd1, mat1, rho1, bgd2, mat2, rho2), df in rates_hybrids.groupby(['parent1_background', 'parent1_mat', 'parent1_rho',\n",
    "                                                                      'parent2_background', 'parent2_mat', 'parent2_rho']):\n",
    "\n",
    "    if df['plasmid'].unique().shape[0] > 1:\n",
    "        \n",
    "        df = df.set_index('plasmid')\n",
    "        hyb = frozenset([bgd1, bgd2])\n",
    "        \n",
    "        b_rate, b_0025, b_0975, b_upper = df.loc['SpB', ['rate', 'rate_0.025', 'rate_0.975', 'upper_bound']]\n",
    "        c_rate, c_0025, c_0975, c_upper = df.loc['SpC', ['rate', 'rate_0.025', 'rate_0.975', 'upper_bound']]\n",
    "        \n",
    "        idx = tuple(df.iloc[0][['parent1_background', 'parent1_mat', 'parent2_background', 'parent2_mat']])\n",
    "            \n",
    "        hyb_included[hyb] = idx\n",
    "        \n",
    "        if any([b_upper, c_upper]):\n",
    "            mfc = 'w'\n",
    "        else:\n",
    "            mfc = hybrid_colors[hyb]\n",
    "        \n",
    "        ax.scatter(b_rate, c_rate, marker='o', color=mfc, edgecolors=hybrid_colors[hyb], zorder=2)\n",
    "        \n",
    "        if not b_upper:\n",
    "            ax.plot([b_0025, b_0975], [c_rate, c_rate], color='k', zorder=1)\n",
    "        if not c_upper:\n",
    "            ax.plot([b_rate, b_rate], [c_0025, c_0975], color='k', zorder=1)\n",
    "    \n",
    "\n",
    "ax.set_xscale('log')\n",
    "ax.set_xlim(xlim)\n",
    "ax.set_xlabel('Transposition rate (cell$^{-1}$ gen$^{-1}$)\\n'+group_alias['SpB']+' Ty1 variant')\n",
    "\n",
    "ax.set_yscale('log')\n",
    "ax.set_ylim(xlim)\n",
    "ax.set_ylabel('Transposition rate (cell$^{-1}$ gen$^{-1}$)\\n'+group_alias['SpC']+' Ty1 variant')\n",
    "\n",
    "ax.plot(xlim, xlim, zorder=0, color='k', lw=0.5)\n",
    "\n",
    "# Add legends\n",
    "\n",
    "ax = fig.add_axes([0.6, 0, 0.4, 1])\n",
    "ax.axis('off')\n",
    "\n",
    "legend_elms0 = [Line2D([0], [0], color='w', marker='o', mfc=fc, ms=10, label=group_alias[l]) for (l, fc) in zip(['SpA','SpB','SpC'], ['#00CD00','#EE0000', '#00008B'])]\n",
    "legend0 = plt.legend(handles=legend_elms0, title='Haploid background', bbox_to_anchor=[0.1, 0.85], loc='upper left')\n",
    "\n",
    "legend_elms1 = [Line2D([0], [0], color='w', marker='o', mfc=fc, mec='k', ms=8, label=l) for (l, fc) in zip(['No','Yes'], ['k', 'w'])]\n",
    "legend1 = plt.legend(handles=legend_elms1, title='Upper bound', bbox_to_anchor=[0.1, 0.7], loc='upper left')\n",
    "\n",
    "legend_elms2 = [Line2D([0], [0], color='w', marker='o', mfc=hybrid_colors[k], ms=10, \n",
    "                       label=hybrid_labels[idx]) for k,idx in hyb_included.items()]\n",
    "legend2 = plt.legend(handles=legend_elms2, title='Diploid background', bbox_to_anchor=[0.1, 0.5], loc='upper left', ncol=1)\n",
    "\n",
    "for legend in [legend0, legend1, legend2]:\n",
    "    ax.add_artist(legend)\n",
    "\n",
    "fig.text(0.02, 0.96, 'A', size=24, fontweight='bold', font='Arial')\n",
    "fig.text(0.02, 0.46, 'B', size=24, fontweight='bold', font='Arial')\n",
    "    \n",
    "    \n",
    "sns.despine()\n",
    "\n",
    "#for ext in ['svg', 'png']:\n",
    "#    plt.savefig(f'{fig_path}Fig6.{ext}')\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
