{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "87593d97-5144-48b5-8b39-f7a3f3eb3265",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import pysam\n",
    "import pickle as pkl\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.patches import FancyArrow\n",
    "from matplotlib import cm\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.ticker import MultipleLocator\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO\n",
    "from scipy import stats\n",
    "from progressbar import ProgressBar\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from roman import toRoman\n",
    "import itertools\n",
    "from collections import Counter\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d7638efa-3d0b-410b-abdc-13b11a8af737",
   "metadata": {},
   "outputs": [],
   "source": [
    "rc_sns = {'ytick_color':'k', 'xtick_color':'k', 'text_color':'k', 'font.sans-serif':'DejaVu Sans', 'figure.facecolor':(1,1,1,1)}\n",
    "sns.set_style(style='ticks', rc=rc_sns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "151255d3-6c52-49b1-88c1-852542581273",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import tables of strain identities and cross parents\n",
    "nano_strains = pd.read_csv('/mnt/HDD3/lrma/script/nano_strains.csv', index_col=0)\n",
    "cross_parents = pd.read_csv('/mnt/HDD3/lrma/private_variants/cross_parents.txt', sep=';', header=None, index_col=0, squeeze=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e6cc1638-5aff-4d6a-a268-35f7cdf58967",
   "metadata": {},
   "outputs": [],
   "source": [
    "cross_alias = {'VL3':'CC1',\n",
    "              'VL4':'CC2',\n",
    "              'VL5':'CC3',\n",
    "              'VL1':'BB1',\n",
    "              'VL2':'BB2',\n",
    "              'L1':'BC1',\n",
    "              'L2':'BC2',\n",
    "              'M1':'BA1',\n",
    "              'M2':'BA2',\n",
    "              'H1':'BSc1',\n",
    "              'H2':'BSc2'}\n",
    "cross_order = {j:i for i,j in enumerate(['CC1', 'CC2', 'CC3', 'BB1', 'BB2', 'BC1', 'BC2', 'BA1', 'BA2', 'BSc1', 'BSc2'])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b29bafd-efe8-4a05-beef-3a396523c479",
   "metadata": {},
   "outputs": [],
   "source": [
    "#split table per subgenome\n",
    "ns_subg = nano_strains.loc[nano_strains['cross']!='P'].copy()\n",
    "for c, df in ns_subg.groupby('cross'):\n",
    "    s1, s2 = cross_parents.loc[c].split(',')\n",
    "    ns_subg.loc[df.index, 's1'] = s1\n",
    "    ns_subg.loc[df.index, 's2'] = s2\n",
    "\n",
    "ns_subg = pd.melt(ns_subg, id_vars=ns_subg.columns[:-2], value_vars=['s1','s2'], value_name='subg', var_name='subg_idx').sort_values(by=['cross','strain','subg_idx'])\n",
    "ns_subg['s_subg'] = ns_subg.apply(lambda x: f'{x[\"strain\"]}.{x[\"subg\"]}', axis=1)\n",
    "ns_subg.index = ns_subg['s_subg'].values\n",
    "# add reference genome for quick mapping \n",
    "for sg, df in ns_subg.groupby('subg'):\n",
    "    if sg in ['LL2013_040','LL2013_054']:\n",
    "        ns_subg['quick_map'] = '/home/mathieu/mhenault_landrylab/Sequences/ref_genomes/S288C_pacbio/S288c.genome.fa'\n",
    "    if sg in ['YPS644','YPS744']:\n",
    "        ns_subg['quick_map'] = '/home/mathieu/mhenault_landrylab/Sequences/ref_genomes/CBS432_pacbio/CBS432.genome.fa'\n",
    "    else:\n",
    "        ns_subg['quick_map'] = '/home/mathieu/mhenault_landrylab/Sequences/ref_genomes/YPS138_pacbio/YPS138.genome.fa'\n",
    "#ns_subg.to_csv('/mnt/HDD3/lrma/script/ns_subg.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "af9e82c7-f710-40f5-bf71-658b0b008807",
   "metadata": {},
   "outputs": [],
   "source": [
    "ns_subg = pd.read_csv('/mnt/HDD3/lrma/script/ns_subg.csv', index_col=0)\n",
    "ns_subg['cross'] = ns_subg['cross'].replace(cross_alias)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cedddf4-7c0a-4c2b-a091-d40399cb8a70",
   "metadata": {},
   "outputs": [],
   "source": [
    "SORT = {}\n",
    "for s in ['J17','C20']:\n",
    "    cross = nano_strains.loc[s, 'cross']\n",
    "    s1, s2 = cross_parents.loc[cross].split(',')\n",
    "\n",
    "    bam = pysam.AlignmentFile(f'/mnt/HDD3/lrma/minimap_sort/{s}.sort.minimap.bam', 'rb')\n",
    "\n",
    "    read_len = {}\n",
    "    query_len = {}\n",
    "    qual = {}\n",
    "    secondary = {}\n",
    "    nm = {}\n",
    "    aln_pos = {}\n",
    "    for i in bam:\n",
    "\n",
    "        name = i.query_name\n",
    "        read_len[name] = i.infer_read_length()\n",
    "        query_len[name] = i.infer_query_length()\n",
    "        secondary[name] = i.is_secondary\n",
    "        aln_pos[name] = i.reference_start\n",
    "        if i.is_secondary:\n",
    "            qual[name] = -1\n",
    "        else:\n",
    "            qual[name] = np.mean(i.query_alignment_qualities)\n",
    "\n",
    "        tags = dict(i.get_tags())\n",
    "        if 'NM' in tags:\n",
    "            nm[name] = tags['NM']\n",
    "        else:\n",
    "            nm[name] = -1\n",
    "\n",
    "    tab = pd.read_csv(f'/mnt/HDD3/lrma/private_variants/private_variants_{cross}.tab', sep='\\t', header=None,\n",
    "                     dtype={0:str,1:np.int32,2:str,3:str,4:str,5:str,6:str})\n",
    "    for gt_col in (5,6):\n",
    "        tab[gt_col] = np.where(tab[gt_col].isin(['1|1','1/1']), tab[4], tab[3])\n",
    "    tab.index = tab[1].values-1\n",
    "\n",
    "    reads = {}\n",
    "    #s1, s2 = samples\n",
    "    idx = 0\n",
    "    with ProgressBar(max_value=tab.shape[0]) as bar:\n",
    "        for tig, tab_sub in tab.groupby(0):\n",
    "            # get list of positions on the contig\n",
    "            loci = bam.pileup(tig)\n",
    "            #iterate pileups per position\n",
    "            for pileup in loci:\n",
    "                i = pileup.reference_pos\n",
    "                # if position is a snp that allows to discriminate parents, continue\n",
    "                if i in tab_sub.index:\n",
    "                    gt1, gt2 = tab_sub.loc[i, [5,6]]\n",
    "\n",
    "                    for (rid, nt) in zip(pileup.get_query_names(), pileup.get_query_sequences()):\n",
    "                        # init reads dict\n",
    "                        if rid not in reads:\n",
    "                            reads[rid] = {s1:0, s2:0}\n",
    "                        nt = nt.upper()\n",
    "                        if nt == gt1:\n",
    "                            reads[rid][s1] += 1\n",
    "                        elif nt == gt2:\n",
    "                            reads[rid][s2] += 1\n",
    "                    idx += 1\n",
    "                    bar.update(idx)\n",
    "\n",
    "    sort = pd.DataFrame(reads).T\n",
    "    #sort['binom'] = sort.apply(lambda x: stats.binom_test(x[s1], x[s1]+x[s2], 0.5), axis=1)\n",
    "    sort['sort'] = sort.apply(lambda x : sort_counts(x, s1, s2), axis=1)\n",
    "\n",
    "    sort['rl'] = pd.Series(read_len).loc[sort.index]\n",
    "    sort['log_rl'] = np.log10(sort['rl'])\n",
    "    sort['ql'] = pd.Series(query_len)\n",
    "    sort['log_ql'] = np.log10(sort['ql'])\n",
    "    sort['sec'] = pd.Series(secondary)\n",
    "    sort['nm'] = pd.Series(nm)\n",
    "    sort['qual'] = pd.Series(qual)\n",
    "    sort['pos'] = pd.Series(aln_pos)\n",
    "    sort['log0'] = np.log10(sort[s1]+1)\n",
    "    sort['log1'] = np.log10(sort[s2]+1)\n",
    "    sort['%nm'] = sort['nm']/sort['ql']\n",
    "\n",
    "    #sort['corr_pval'] = multipletests(sort['binom'], alpha=0.05, method='fdr_bh')[0]\n",
    "    SORT[s] = sort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7b8f797-bec6-49b4-abe6-e18f3c6cf6a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_counts(x, s1, s2):\n",
    "    c1, c2 = x[[s1, s2]]\n",
    "    if np.max([c1 ,c2]) >= 2:\n",
    "        if c1 >= 2*c2:\n",
    "            return s1\n",
    "        if c2 >= 2*c1:\n",
    "            return s2\n",
    "        else:\n",
    "            return -1\n",
    "    else:\n",
    "        return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5fd9385-de30-4f18-8367-3a4fb41f8184",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gt, df in sort.loc[sort['corr_pval']].groupby('sort'):\n",
    "    with open(f'/mnt/HDD3/lrma/sort/{s}.{gt}.reads', 'w') as handle:\n",
    "        handle.write('\\n'.join(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76ed3ec8-3205-4ebe-afa3-9fdad2a39b7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(ncols=2, nrows=2, gridspec_kw={'height_ratios':(1,4), 'wspace':0.4}, figsize=[8,5])\n",
    "for (s, sort), ax in zip(SORT.items(), axes.T):\n",
    "    \n",
    "    cross = nano_strains.loc[s, 'cross']\n",
    "    s1, s2 = cross_parents.loc[cross].split(',')\n",
    "    \n",
    "    palette={s1:'blue',s2:'red',-1:'0.3'}\n",
    "    \n",
    "    sns.scatterplot(data=sort, x=s1, y=s2, hue='sort', edgecolor=(1,1,1,0), ax=ax[1],\n",
    "                   palette=palette, alpha=0.1, s=10)\n",
    "    ax[1].set_xlabel(f'snps {s1}')\n",
    "    ax[1].set_ylabel(f'snps {s2}')\n",
    "    \n",
    "    bar = sort.value_counts('sort')\n",
    "    ax[0].bar(range(3), bar, color=[palette[i] for i in bar.index])\n",
    "    ax[0].set_xticks(range(3))\n",
    "    ax[0].set_xticklabels(bar.index)\n",
    "    ax[0].set_ylabel('sorted reads')\n",
    "    ax[0].set_title(f'{s} ({cross_alias[cross]})')\n",
    "\n",
    "sns.despine()\n",
    "plt.savefig('/mnt/HDD3/lrma/fig/sort_example.png', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3df4c2f4-4bb7-42dd-8d9d-0c79d03f7d88",
   "metadata": {},
   "source": [
    "### correlate numbers of reads in mappings to sorted subgenomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6af541a-e851-4edc-9461-7a3cdfc4a325",
   "metadata": {},
   "outputs": [],
   "source": [
    "#parse mapping stats\n",
    "map_sort_stats = {}\n",
    "for s in nano_strains.loc[nano_strains['cross']!='P'].index:\n",
    "    with open(f'/mnt/HDD3/lrma/minimap_sort/{s}.sort.minimap.stats') as handle:\n",
    "        dat = dict([line.split('\\t')[1:3] for line in handle.read().splitlines() if line[:2]=='SN'])\n",
    "        for m,v in dat.items():\n",
    "            if '.' in v:\n",
    "                dat[m] = float(v)\n",
    "            else:\n",
    "                dat[m] = int(v)\n",
    "        map_sort_stats[s] = dat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f90628b5-46e6-4ee7-be59-bc1bca6b68b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#parse number of bases\n",
    "read_lengths = {}\n",
    "idx = 0\n",
    "with ProgressBar(max_value=ns_subg.shape[0]) as bar:\n",
    "    for s in ns_subg.index:\n",
    "        read_lengths[s]  = np.array([len(seq.seq) for seq in SeqIO.parse(f'/mnt/HDD3/lrma/sort/{s}.reads.fastq', 'fastq')])\n",
    "        idx += 1\n",
    "        bar.update(idx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "403d7d0a-c713-4ed8-a25d-826ed969932f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#with open('/mnt/HDD3/lrma/sort/rl.pkl', 'wb') as handle:\n",
    "#    pkl.dump(read_lengths, handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3469f9b4-24d7-4eb3-9464-31c17f853798",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/mnt/HDD3/lrma/sort/rl.pkl', 'rb') as handle:\n",
    "    read_lengths = pkl.load(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04d63a5f-2b4b-45b7-8e00-f08b8909ecb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#parse read counts from sorted subgenomes\n",
    "for s in ns_subg.index:\n",
    "    ns_subg.loc[s, 'bases_subg'] = read_lengths[s].sum()\n",
    "    ns_subg.loc[s, 'n_reads_subg'] = read_lengths[s].shape[0]\n",
    "for s, df in ns_subg.groupby('strain'):\n",
    "    ns_subg.loc[df.index, 'n_reads_map'] = map_sort_stats[s]['reads mapped:']\n",
    "    ns_subg.loc[df.index, 'bases_map'] = map_sort_stats[s]['bases mapped (cigar):']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44c94bff-f29c-48ac-9085-e5ac076e3dc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot\n",
    "fig, axes = plt.subplots(ncols=2, figsize=[8,4])\n",
    "lw_ploid = {True:1, False:2}\n",
    "\n",
    "for ax, m1, m2, a1, a2 in zip(axes, ['n_reads_map','bases_map'], ['n_reads_subg','bases_subg'],\n",
    "                              ['total reads', 'total bases'], ['subgenome reads', 'subgenome bases']):\n",
    "\n",
    "    for (s,c,d), df in ns_subg.loc[ns_subg['identity_filter']].groupby(['strain','cross_color','diploid_filter']):\n",
    "        ax.plot(df[m1], df[m2], c=c, lw=lw_ploid[d])\n",
    "\n",
    "    #ax.set_xscale('log')\n",
    "    #ax.set_yscale('log')\n",
    "        \n",
    "    xlim = np.array(ax.get_xlim())\n",
    "    for p, ls in zip([0.5, 0.333, 0.667], ['-',':',':']):\n",
    "        ax.plot(xlim, p*xlim, zorder=0, lw=0.5, c='0.5', ls=ls)\n",
    "        \n",
    "    ax.set_xlabel(a1)\n",
    "    ax.set_ylabel(a2)\n",
    "    \n",
    "plt.tight_layout()\n",
    "sns.despine()\n",
    "#plt.savefig('/mnt/HDD3/lrma/fig/coverage_subgenome.png', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9b1b43a-adc3-4b3d-b2fc-6b66dbc2e280",
   "metadata": {},
   "outputs": [],
   "source": [
    "dat = ns_subg.sort_values('bases_subg').groupby('strain').apply(lambda x: x.iloc[0])['bases_subg']\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "ax.plot(np.sort(dat), np.linspace(0,1,dat.shape[0]))\n",
    "ax.axvline(15*1.2e7)\n",
    "\n",
    "ax.axvline(dat.loc['C12'], color='red')\n",
    "\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1c4559c-e509-4c17-83b4-a34cd74c9a6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get mapping positions per subgenome\n",
    "# ref genome used for sort mappings\n",
    "genome = {}\n",
    "tig_offsets = []\n",
    "for seq in SeqIO.parse('/home/mathieu/paradoxus_nanopore/paradoxus4/assemblies/paradoxus4_bc02_smartdenovo_pilon-nanopolish_reordered.fasta', 'fasta'):\n",
    "    genome[seq.id] = seq\n",
    "    tig_offsets.append([seq.id, len(seq.seq)])\n",
    "tig_offsets = pd.DataFrame(tig_offsets)\n",
    "tig_offsets['off'] = [0] + list(np.cumsum(tig_offsets[1])[:-1])\n",
    "tig_offsets['end'] = tig_offsets['off']+tig_offsets[1]-1\n",
    "tig_offsets.index = tig_offsets[0].values\n",
    "\n",
    "tig_chrom = pd.read_csv('/home/mathieu/paradoxus_nanopore/paradoxus4/barcode02/smartdenovo/mauve/tig_rearrangement.txt', sep='\\t', header=None)\n",
    "for chrom, df in tig_chrom.groupby(0):\n",
    "    if int(chrom[3:])%2 == 1:\n",
    "        tig_chrom.loc[df.index, 'color'] = '0'\n",
    "    if int(chrom[3:])%2 == 0:\n",
    "        tig_chrom.loc[df.index, 'color'] = '0.4'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6129ee0-9538-4b1c-bc9f-b6fb5ccd5542",
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse read pos\n",
    "POS = {}\n",
    "window = 1e5\n",
    "for s in ns_subg.index:\n",
    "    dat = pd.read_csv(f'/mnt/HDD3/lrma/sort/{s}.pos.txt', sep='\\t', header=None)\n",
    "    #add offsets\n",
    "    dat[1] = dat[1] + tig_offsets.loc[dat[0], 'off'].values\n",
    "    dat = np.histogram(dat[1], bins=np.arange(0, 1.2e7, window))\n",
    "    dat = pd.Series(dat[0], name=s)\n",
    "    dat = dat/np.median(dat)\n",
    "    POS[s] = dat\n",
    "\n",
    "POS = pd.concat(POS, axis=1).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68bfc18b-9333-4f62-a045-40e144ed2907",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for cross, df in ns_subg.loc[ns_subg['identity_filter']].sort_values(by=['strain','subg_idx']).groupby('cross'):\n",
    "    #if cross == 'H1':\n",
    "    fig, ax = plt.subplots()\n",
    "    sns.heatmap(POS.loc[df.index], cmap='coolwarm', vmin=0, vmax=4, center=1)\n",
    "\n",
    "    for (chrom, c), df1 in tig_chrom.groupby([0,'color']):\n",
    "        df1 = tig_offsets.loc[df1[1].apply(lambda x: f'{x}_pilon')]\n",
    "        start, length = 1/window * np.array([df1.iloc[0]['off'], df1[1].sum()])\n",
    "        ar = FancyArrow(start, df.shape[0]+2, length, 0, width=0.8, head_width=0.8,\n",
    "                        fc=c, lw=0, length_includes_head=True, clip_on=False, head_length=0)\n",
    "        ax.add_patch(ar)\n",
    "        ax.text(start+0.5*length, df.shape[0]+2, toRoman(int(chrom[3:])), \n",
    "                ha='center', va='center', size=7, color='w')\n",
    "        \n",
    "        ax.axvline(start+length, lw=1, color='white')\n",
    "        for i in np.arange(0, df.shape[0]+1, 2):\n",
    "            ax.axhline(i, lw=1, color='white')\n",
    "\n",
    "    ax.set_title(cross_alias[cross])\n",
    "    ax.set_xticks(np.arange(0,120,10))\n",
    "    ax.set_xticklabels([f'{i}M' for i in np.arange(0,12,1)], rotation=0, size=8)\n",
    "    ax.set_yticks(range(df.shape[0]))\n",
    "    ax.set_yticklabels(df.index, size=8)\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    #plt.savefig(f'/mnt/HDD3/lrma/fig/hm_cov_subgenome_{cross_alias[cross]}.png', dpi=300)\n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68a4ada3-7a4c-499c-9cca-a5d7c3ec8035",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "dat = POS.loc[ns_subg.sort_values(by=['cross','strain','subg_idx']).index]\n",
    "\n",
    "sns.heatmap(dat, cmap='coolwarm', vmin=0, vmax=4, center=1)\n",
    "\n",
    "for (chrom, c), df1 in tig_chrom.groupby([0,'color']):\n",
    "    df1 = tig_offsets.loc[df1[1].apply(lambda x: f'{x}_pilon')]\n",
    "    start, length = 1/window * np.array([df1.iloc[0]['off'], df1[1].sum()])\n",
    "    ar = FancyArrow(start, df.shape[0]+2, length, 0, width=0.8, head_width=0.8,\n",
    "                    fc=c, lw=0, length_includes_head=True, clip_on=False, head_length=0)\n",
    "    ax.add_patch(ar)\n",
    "    ax.text(start+0.5*length, df.shape[0]+2, toRoman(int(chrom[3:])), \n",
    "            ha='center', va='center', size=7, color='w')\n",
    "\n",
    "    ax.axvline(start+length, lw=1, color='white')\n",
    "    for i in np.arange(0, df.shape[0]+1, 2):\n",
    "        ax.axhline(i, lw=1, color='white')\n",
    "\n",
    "#ax.set_title(cross_alias[cross])\n",
    "ax.set_xticks(np.arange(0,120,10))\n",
    "ax.set_xticklabels([f'{i}M' for i in np.arange(0,12,1)], rotation=0, size=8)\n",
    "#ax.set_yticks(range(df.shape[0]))\n",
    "#ax.set_yticklabels(df.index, size=8)\n",
    "ax.set_yticklabels([])\n",
    "\n",
    "plt.tight_layout()\n",
    "#plt.savefig(f'/mnt/HDD3/lrma/fig/hm_cov_subgenome_{cross_alias[cross]}.png', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5ce9c8a-b479-41f3-9612-2c530fb4d68c",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.stat(f'/mnt/HDD3/lrma/wtdbg2/{s}/{s}.cns.fa').st_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8de14070-ed2f-4c3f-b7eb-9905cd7a749f",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = '/mnt/HDD3/lrma/medaka/A30.LL2011_004/consensus.fasta'\n",
    "ref = '/home/mathieu/paradoxus_nanopore/MA_parents/assemblies/LL2011_004.chromosomes.fasta'\n",
    "\n",
    "# get filesize\n",
    "if os.stat(query).st_size > 0:\n",
    "    tig_off = {}\n",
    "    for i, file in enumerate([ref, query]):\n",
    "        tig_off[i] = pd.concat([pd.Series([seq.id, len(seq.seq)]) for seq in SeqIO.parse(file, 'fasta')], axis=1).T\n",
    "        tig_off[i].index = tig_off[i][0].values\n",
    "        # if ref, sort now. if query, sort after\n",
    "        if i==0:\n",
    "            tig_off[i][2] = np.concatenate([np.array([0]), np.cumsum(tig_off[i][1].values)[:-1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57614538-05ea-4d2e-bb55-bf2ca142e1d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "        coords = pd.read_csv(coords, sep='\\t', \n",
    "                             #skiprows=4,\n",
    "                             header=None)\n",
    "        # remove hits shorter than 10kb\n",
    "        coords = coords.loc[(coords[4]>1e4) & (coords[5]>1e4)]\n",
    "\n",
    "        if coords.shape[0] > 0:\n",
    "            for j in (0,1):\n",
    "                coords[j] += tig_off[0].loc[coords[9], 2].values\n",
    "            order_query = coords.groupby(10).apply(lambda x: np.mean(x.sort_values(by=5, ascending=False).iloc[0, [0,1]])).sort_values()\n",
    "            tig_off[1] = tig_off[1].loc[order_query.index]\n",
    "            tig_off[1][2] = np.concatenate([np.array([0]), np.cumsum(tig_off[1][1].values)[:-1]])\n",
    "            for j in (2,3):\n",
    "                coords[j] += tig_off[1].loc[coords[10], 2].values\n",
    "\n",
    "            # flip query contigs\n",
    "            for tig, df in coords.groupby(10):\n",
    "                if (df[8]*df[5]).mean() < 0:\n",
    "                    coords_copy = df[[2,3]].copy()\n",
    "                    midpoint = np.mean([(coords_copy[3].min(), coords_copy[2].max())])\n",
    "                    coords.loc[df.index, [2,3]] = -1*(coords_copy-midpoint)+midpoint\n",
    "                    coords.loc[df.index, 8] *= -1\n",
    "\n",
    "            for i in coords.index:\n",
    "                if coords.loc[i, 8] == -1:\n",
    "                    c = 'red'\n",
    "                else:\n",
    "                    c = 'black'\n",
    "\n",
    "                ax.plot(coords.loc[i, [0,1]], coords.loc[i, [2,3]], c=c, lw=2, alpha=0.5)\n",
    "\n",
    "\n",
    "            if decorate in ('ref','both'):\n",
    "                for tig in tig_off[0].index:\n",
    "                    pos = tig_off[0].loc[tig, 2]\n",
    "                    ax.axvline(pos, lw=0.5, ls=':', color='0.5')\n",
    "                    if tig_off[0].loc[tig, 1] > 1e5:\n",
    "                        ax.text(pos, 0, tig, ha='left', va='top', size=7, rotation=90)\n",
    "            if decorate == 'both':\n",
    "                for tig in tig_off[1].index:\n",
    "                    pos = tig_off[1].loc[tig, 2]\n",
    "                    ax.axhline(pos, lw=0.5, ls=':', color='0.5')\n",
    "                    if tig_off[1].loc[tig, 1] > 1e5:\n",
    "                        ax.text(0, pos, tig, ha='right', va='bottom', size=7)\n",
    "                        \n",
    "            if return_tig_off:\n",
    "                return tig_off"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "2f77891c-f859-4124-b47a-1923394ec82b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_mummer(s, ref, query, coords, ax, decorate=False, return_tig_off=False):  \n",
    "\n",
    "    # get filesize\n",
    "    if os.stat(query).st_size > 0:\n",
    "        tig_off = {}\n",
    "        for i, file in enumerate([ref, query]):\n",
    "            tig_off[i] = pd.concat([pd.Series([seq.id, len(seq.seq)]) for seq in SeqIO.parse(file, 'fasta')], axis=1).T\n",
    "            tig_off[i].index = tig_off[i][0].values\n",
    "            # if ref, sort now. if query, sort after\n",
    "            if i==0:\n",
    "                tig_off[i][2] = np.concatenate([np.array([0]), np.cumsum(tig_off[i][1].values)[:-1]])\n",
    "\n",
    "        coords = pd.read_csv(coords, sep='\\t', \n",
    "                             #skiprows=4,\n",
    "                             header=None)\n",
    "        # remove hits shorter than 10kb\n",
    "        coords = coords.loc[(coords[4]>1e4) & (coords[5]>1e4)]\n",
    "\n",
    "        if coords.shape[0] > 0:\n",
    "            for j in (0,1):\n",
    "                coords[j] += tig_off[0].loc[coords[9], 2].values\n",
    "            order_query = coords.groupby(10).apply(lambda x: np.mean(x.sort_values(by=5, ascending=False).iloc[0, [0,1]])).sort_values()\n",
    "            tig_off[1] = tig_off[1].loc[order_query.index]\n",
    "            tig_off[1][2] = np.concatenate([np.array([0]), np.cumsum(tig_off[1][1].values)[:-1]])\n",
    "            for j in (2,3):\n",
    "                coords[j] += tig_off[1].loc[coords[10], 2].values\n",
    "\n",
    "            # flip query contigs\n",
    "            for tig, df in coords.groupby(10):\n",
    "                if (df[8]*df[5]).mean() < 0:\n",
    "                    coords_copy = df[[2,3]].copy()\n",
    "                    midpoint = np.mean([(coords_copy[3].min(), coords_copy[2].max())])\n",
    "                    coords.loc[df.index, [2,3]] = -1*(coords_copy-midpoint)+midpoint\n",
    "                    coords.loc[df.index, 8] *= -1\n",
    "\n",
    "            for i in coords.index:\n",
    "                if coords.loc[i, 8] == -1:\n",
    "                    c = 'red'\n",
    "                else:\n",
    "                    c = 'black'\n",
    "\n",
    "                ax.plot(coords.loc[i, [0,1]], coords.loc[i, [2,3]], c=c, lw=1, alpha=0.5)\n",
    "\n",
    "\n",
    "            if decorate in ('ref','both'):\n",
    "                for tig in tig_off[0].index:\n",
    "                    pos = tig_off[0].loc[tig, 2]\n",
    "                    ax.axvline(pos, lw=1, ls=':', color='blue', zorder=0)\n",
    "                    if tig_off[0].loc[tig, 1] > 1e5:\n",
    "                        ax.text(pos, 0, tig, c='blue', ha='left', va='top', size=7, rotation=90)\n",
    "            if decorate == 'both':\n",
    "                for tig in tig_off[1].index:\n",
    "                    pos = tig_off[1].loc[tig, 2]\n",
    "                    ax.axhline(pos, lw=1, ls=':', color='blue', zorder=0)\n",
    "                    if tig_off[1].loc[tig, 1] > 1e5:\n",
    "                        ax.text(0, pos, tig, c='blue', ha='right', va='bottom', size=7)\n",
    "                        \n",
    "            if return_tig_off:\n",
    "                return tig_off"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a4643d8-a724-45bd-bdc0-3fee0439b7f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#quick alignment of draft assemblies\n",
    "\n",
    "for (cross, subg), df in ns_subg.loc[ns_subg['identity_filter']].groupby(['cross','subg']):\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=[7,7])\n",
    "    \n",
    "    decorate = True\n",
    "    \n",
    "    for (s,r), df1 in df.groupby(['s_subg','subg']):\n",
    "        ref = f'/home/mathieu/paradoxus_nanopore/MA_parents/assemblies/{r}.chromosomes.fasta'\n",
    "        qry = f'/mnt/HDD3/lrma/chroder/{s}.chroder.qry.fasta'\n",
    "        coords = f'/mnt/HDD3/lrma/chroder/{s}.chroder.coords'\n",
    "        plot_mummer(s, ref, qry, coords, ax, decorate=decorate)\n",
    "        decorate = False\n",
    "\n",
    "    ax.set_title(f'{cross} {subg}')\n",
    "    ax.set_xlabel('reference genome')\n",
    "    ax.set_ylabel('draft genome')\n",
    "    \n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f'/mnt/HDD3/lrma/fig/draft_assemblies_{cross}_{subg}.png', dpi=300)\n",
    "    #plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6a5bb344-14ff-4a93-b823-8bfa0eb2c1ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(range(10), range(10), color='red')\n",
    "plt.savefig('/mnt/HDD1/Dropbox/temp_sync/range10.png', dpi=600)\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "7e32331d-3a8e-4060-be8b-ebe9673e8456",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(254, 18)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ns_subg.loc[ns_subg['identity_filter']].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "726f1957-0549-47eb-a2f6-514362cc9ed7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100% (254 of 254) |######################| Elapsed Time: 0:06:22 Time:  0:06:22\n"
     ]
    }
   ],
   "source": [
    "#quick alignment of draft assemblies, per-line\n",
    "\n",
    "idx_bar = 0\n",
    "with ProgressBar(max_value=254) as bar:\n",
    "    for (s, cross, ref), df in ns_subg.loc[ns_subg['identity_filter']].groupby(['s_subg','cross','ref_chroder']):\n",
    "    #if s == 'B7.MSH-604':\n",
    "        qry = f'/mnt/HDD3/lrma/chroder/{s}.chroder.qry.fasta'\n",
    "        coords = f'/mnt/HDD3/lrma/chroder/{s}.chroder.fil.coords'\n",
    "\n",
    "        if os.path.exists(qry) and os.path.exists(coords):\n",
    "\n",
    "            fig, ax = plt.subplots(figsize=[7,7])\n",
    "            plot_mummer(s, ref, qry, coords, ax, decorate='both')\n",
    "\n",
    "            ax.set_title(f'{s}')\n",
    "            ax.set_xlabel(ref.split('/')[-1])\n",
    "            ax.set_ylabel(s)\n",
    "            \n",
    "            ax.set_xticks(np.arange(0, 1.2e7, 1e6))\n",
    "            ax.xaxis.set_minor_locator(MultipleLocator(1e5))\n",
    "            #ax.yaxis.set_minor_locator(MultipleLocator(1e5))\n",
    "            ax.grid(axis='x', which='minor', lw=0.3)\n",
    "            ax.grid(axis='x', which='major', lw=0.7)\n",
    "            #ax.grid(axis='y', which='minor', lw=0.3)\n",
    "            #ax.grid(axis='y', which='major', lw=0.7)\n",
    "            \n",
    "\n",
    "            sns.despine()\n",
    "            plt.tight_layout()\n",
    "            plt.savefig(f'/mnt/HDD3/lrma/fig/draft_assemblies/draft_assembly_{cross}_{s}.png', dpi=300)\n",
    "            #plt.savefig(f'/mnt/HDD1/Dropbox/temp_sync/draft_assembly_{cross}_{s}.png', dpi=500)\n",
    "            plt.close()\n",
    "            \n",
    "        idx_bar += 1\n",
    "        bar.update(idx_bar)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d96fe33f-3fb4-49ac-b444-cc991d61811f",
   "metadata": {},
   "source": [
    "# dnadiff analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49593efc-40cf-4938-b9a3-660988764693",
   "metadata": {},
   "outputs": [],
   "source": [
    "cross_color = ns_subg.groupby('cross').apply(lambda x: x.iloc[0]['cross_color']).to_dict()\n",
    "subg_order = ['MSH-604','UWOPS-91-202','LL2012_021','LL2012_028',\n",
    "             'LL2011_004','LL2011_009','MSH-587-1','LL2011_012','LL2011_001',\n",
    "             'YPS644', 'YPS744', 'LL2013_040', 'LL2013_054']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebef2ced-e937-4308-8b9a-e4816f4b0dec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse mummer dnadiff report\n",
    "dnadiff = []\n",
    "for s in ns_subg.loc[ns_subg['identity_filter']].index:\n",
    "    dat = pd.read_csv(f'/mnt/HDD3/lrma/medaka/{s}/{s}.report', skiprows=4, header=None, sep=' +', comment='[').dropna(how='all')\n",
    "    dat = dat.drop(range(37,50))\n",
    "    dat = dat.drop(range(59,68))\n",
    "    dat.columns = ['feat', '', 'value']\n",
    "    dat['s_subg'] = s\n",
    "    dnadiff.append(dat.iloc[:, [0,2,3]])\n",
    "dnadiff = pd.concat(dnadiff).reset_index()\n",
    "dnadiff = pd.concat([dnadiff, dnadiff['value'].apply(lambda x: pd.Series(re.match('(\\d+)(?:\\((\\d+\\.\\d+)%\\))*', x).groups(), index=['count','perc']))], axis=1)\n",
    "dnadiff['count'] = dnadiff['count'].astype(int)\n",
    "dnadiff['perc'] = dnadiff['perc'].astype(float)\n",
    "for c in ['strain', 'subg', 'cross', 'batch']:\n",
    "    dnadiff[c] = ns_subg.loc[dnadiff['s_subg'], c].values\n",
    "\n",
    "subst_dict = {'AT':'A_T',\n",
    "             'AC':'A_C',\n",
    "             'AG':'A_G',\n",
    "             'CA':'C_A',\n",
    "             'CT':'C_T',\n",
    "             'CG':'C_G',\n",
    "             'TA':'A_T',\n",
    "             'TC':'A_G',\n",
    "             'TG':'A_C',\n",
    "             'GA':'C_T',\n",
    "             'GT':'C_A',\n",
    "             'GC':'C_G'}\n",
    "indel_dict = {'.A':'.A',\n",
    "             '.T':'.A',\n",
    "             '.C':'.C',\n",
    "             '.G':'.C',\n",
    "             'A.':'A.',\n",
    "             'T.':'A.',\n",
    "             'C.':'C.',\n",
    "             'G.':'C.'}\n",
    "for f, df in dnadiff.groupby('feat'):\n",
    "    if len(f) == 2:\n",
    "        if re.match('[ATCG]\\.', f) or re.match('\\.[ATCG]', f):\n",
    "            dnadiff.loc[df.index, 'feat_type'] = 'indel'\n",
    "            dnadiff.loc[df.index, 'indel_type'] = indel_dict[f]\n",
    "        if re.match('[ATCG]{2}', f):\n",
    "            dnadiff.loc[df.index, 'feat_type'] = 'subst'\n",
    "            dnadiff.loc[df.index, 'subst_type'] = subst_dict[f]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aabcf869-008b-403b-a26b-7ad65ae3bf36",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = dnadiff.loc[dnadiff['feat_type']=='subst']\n",
    "fig, ax = plt.subplots(figsize=[12,5])\n",
    "sns.boxplot(x='cross', y='count', hue='subst_type', order=cross_order,\n",
    "            data=data)\n",
    "ax.set_yscale('log')\n",
    "ax.legend(loc=6, frameon=False, bbox_to_anchor=[1,0.5])\n",
    "\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.savefig('/mnt/HDD3/lrma/fig/subst_medaka_cross.png', dpi=300)\n",
    "plt.show()\n",
    "plt.close()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[9,4])\n",
    "sns.boxplot(x='batch', y='count', hue='subst_type', order=[f'batch{i}' for i in range(1,8)],\n",
    "            data=data)\n",
    "ax.set_yscale('log')\n",
    "ax.legend(loc=6, frameon=False, bbox_to_anchor=[1,0.5])\n",
    "\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.savefig('/mnt/HDD3/lrma/fig/subst_medaka_batch.png', dpi=300)\n",
    "plt.show()\n",
    "plt.close()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,5])\n",
    "sns.boxplot(x='subg', y='count', hue='subst_type', order=subg_order,\n",
    "            data=data)\n",
    "ax.set_yscale('log')\n",
    "ax.set_xticklabels(subg_order, rotation=30)\n",
    "ax.legend(loc=6, frameon=False, bbox_to_anchor=[1,0.5])\n",
    "\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.savefig('/mnt/HDD3/lrma/fig/subst_medaka_subg.png', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f02cdb2-d35a-46e8-8f9e-99865998af1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = dnadiff.loc[dnadiff['feat_type']=='indel']\n",
    "fig, ax = plt.subplots(figsize=[12,5])\n",
    "sns.boxplot(x='cross', y='count', hue='indel_type', order=cross_order,\n",
    "            data=data)\n",
    "ax.set_yscale('log')\n",
    "ax.legend(loc=6, frameon=False, bbox_to_anchor=[1,0.5])\n",
    "\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.savefig('/mnt/HDD3/lrma/fig/indel_medaka_cross.png', dpi=300)\n",
    "plt.show()\n",
    "plt.close()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[9,4])\n",
    "sns.boxplot(x='batch', y='count', hue='indel_type', order=[f'batch{i}' for i in range(1,8)],\n",
    "            data=data)\n",
    "ax.set_yscale('log')\n",
    "ax.legend(loc=6, frameon=False, bbox_to_anchor=[1,0.5])\n",
    "\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.savefig('/mnt/HDD3/lrma/fig/indel_medaka_batch.png', dpi=300)\n",
    "plt.show()\n",
    "plt.close()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,5])\n",
    "sns.boxplot(x='subg', y='count', hue='indel_type', order=subg_order,\n",
    "            data=data)\n",
    "ax.set_yscale('log')\n",
    "ax.set_xticklabels(subg_order, rotation=30)\n",
    "ax.legend(loc=6, frameon=False, bbox_to_anchor=[1,0.5])\n",
    "\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.savefig('/mnt/HDD3/lrma/fig/indel_medaka_subg.png', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e57c77e-afd6-41dc-aa55-1d5457a80334",
   "metadata": {},
   "outputs": [],
   "source": [
    "SYRI['<DEL>']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e13af56e-d8c5-47ea-bc78-90ee51aad0b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "SYRI = []\n",
    "\n",
    "for s in ns_subg.loc[ns_subg['identity_filter']].index:\n",
    "\n",
    "    vcf_path = f'/mnt/HDD3/lrma/syri/{s}/{s}.syri.vcf'\n",
    "    if os.stat(vcf_path).st_size > 0:\n",
    "        vcf = pd.read_csv(vcf_path, comment='#', header=None, sep='\\t')\n",
    "        vcf = vcf.rename(columns={4:'4'})\n",
    "        vcf = vcf.value_counts('4').reset_index()\n",
    "        vcf.columns = ['sv_type','count']\n",
    "        vcf['s_subg'] = s\n",
    "        SYRI.append(vcf)\n",
    "\n",
    "SYRI = pd.concat(SYRI).reset_index(drop=True)\n",
    "SYRI['sv_type'] = SYRI['sv_type'].apply(lambda x: x.strip('<>'))\n",
    "for c in ['strain', 'subg', 'cross', 'batch']:\n",
    "    SYRI[c] = ns_subg.loc[SYRI['s_subg'], c].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c83363ac-9613-4fd7-a900-c327b95e3b33",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = SYRI.loc[SYRI['sv_type'].isin(['DUP','INVDP','TRANS','INVTR'])]\n",
    "fig, axes = plt.subplots(ncols=2, nrows=2, figsize=[9,6])\n",
    "\n",
    "for (ax_y, ax_x), (sv, df) in zip(itertools.product(range(2), range(2)), data.groupby('sv_type')):\n",
    "    ax = axes[ax_y, ax_x]\n",
    "    sns.boxplot(x='cross', y='count', data=df, ax=ax, palette=cross_color, order=cross_order)\n",
    "    ax.set_title(sv)\n",
    "    \n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.savefig('/mnt/HDD3/lrma/fig/syri_sv_counts.png', dpi=300)\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47eae947-e437-4f58-90bc-51b16d00c11c",
   "metadata": {},
   "outputs": [],
   "source": [
    "r = 'LL2012_028'\n",
    "s = 'I8.LL2012_028'\n",
    "\n",
    "ref = f'/home/mathieu/paradoxus_nanopore/MA_parents/assemblies/{r}.chromosomes.fasta'\n",
    "qry = f'/mnt/HDD3/lrma/chroder/{s}.chroder.qry.fasta'\n",
    "coords = f'/mnt/HDD3/lrma/chroder/{s}.chroder.fil.coords'\n",
    "\n",
    "#vcf = pd.read_csv('/mnt/HDD3/lrma/syri/testsyri.vcf', comment='#', header=None, sep='\\t')\n",
    "vcf = pd.read_csv('/mnt/HDD3/lrma/syri/I8.LL2012_028.syri.vcf', comment='#', header=None, sep='\\t')\n",
    "vcf = pd.concat([vcf, vcf[7].apply(lambda x: pd.Series(dict([i.split('=') for i in x.split(';')])))], axis=1)\n",
    "vcf = vcf.replace('.', np.nan).astype({'END':np.float32,'StartB':np.float32,'EndB':np.float32})\n",
    "# only keep vcf entries for which a chrB is defined \n",
    "vcf = vcf.loc[vcf['ChrB'].apply(lambda x: type(x)==str)]\n",
    "for col in [1, 'END']:\n",
    "    vcf[col] = vcf[col] + tig_off[0].loc[vcf[0], 2].values\n",
    "for col in ['StartB', 'EndB']:\n",
    "    vcf[col] = vcf[col] + tig_off[1].loc[vcf['ChrB'], 2].values\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[7,6])\n",
    "tig_off = plot_mummer(s, ref, qry, coords, ax, decorate='both', return_tig_off=True)\n",
    "    \n",
    "sv_color = dict(zip(['<DUP>','<INVDP>','<TRANS>','<INVTR>'], [cm.tab10(i) for i in range(4)]))\n",
    "    \n",
    "for sv, df in vcf.loc[vcf[4].isin(['<DUP>','<INVDP>','<TRANS>','<INVTR>'])].groupby(4):\n",
    "    \n",
    "    \n",
    "    for i in df.index:\n",
    "        x = vcf.loc[i, [1,'END']].astype(float)\n",
    "        y = vcf.loc[i, ['StartB','EndB']].astype(float)\n",
    "        ax.plot(x, y, lw=1, marker='o', ms=4, color=sv_color[sv])\n",
    "\n",
    "ax.set_xlabel(ref.split('/')[-1])\n",
    "ax.set_ylabel(s)\n",
    "ax.legend(handles=[Line2D([0], [0], color='w', marker='s', mfc=c, label=l.strip('<>')) for l,c in sv_color.items()],\n",
    "         loc=6, bbox_to_anchor=[1,0.5], frameon=False)\n",
    "\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.savefig('/mnt/HDD3/lrma/fig/syri_I8.LL2012_028.dotplot.png', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d0c43ea-0f83-405d-8ba9-555bd11445a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "set(vcf[4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "820b8e85-01dc-46c3-a4ab-c8d2fa8201aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get and plot stats from the draft assemblies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ee51903-4399-4b92-a437-27ddfaec8942",
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_size = {}\n",
    "for subg in set(ns_subg['subg']):\n",
    "    file = f'/home/mathieu/paradoxus_nanopore/MA_parents/assemblies/{subg}.chromosomes.fasta'\n",
    "    parent_size[subg] = np.sum([len(seq.seq) for seq in SeqIO.parse(file, 'fasta')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1feda4e3-a32e-4489-9d3b-35faaa22235b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get statistics on assemblies\n",
    "assembl_stats = []\n",
    "for (cross, subg, s), df in ns_subg.loc[ns_subg['identity_filter']].groupby(['cross','subg','s_subg']):\n",
    "    cross = cross_alias[cross]\n",
    "    query = f'/mnt/HDD3/lrma/medaka/{s}/consensus.fasta'\n",
    "    \n",
    "    draft = pd.concat([pd.Series([seq.id, len(seq.seq)]) for seq in SeqIO.parse(query, 'fasta')], axis=1).T.sort_values(by=1, ascending=False)\n",
    "    Size = draft[1].sum()\n",
    "    n = draft.shape[0]\n",
    "    idx, cum = (0,0)\n",
    "    while cum < Size/2:\n",
    "        tig, size = draft.loc[idx]\n",
    "        cum += size\n",
    "        idx += 1\n",
    "        \n",
    "    N50 = size\n",
    "    assembl_stats.append([cross, cross_order[cross], subg, f'{cross}.{subg}', s, Size, N50, n])\n",
    "\n",
    "assembl_stats = pd.DataFrame(assembl_stats, columns=['cross','co','subg','cross_subg','s_subg','size','N50','n'])\n",
    "assembl_stats.index = assembl_stats['s_subg'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a789bed-e48f-4065-9f2b-f5df118fdd9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subg, df in assembl_stats.groupby('subg'):\n",
    "    assembl_stats.loc[df.index, 'size_ratio'] = df['size']/parent_size[subg]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f483caa8-c3ea-4264-9d0e-7935cb8e3809",
   "metadata": {},
   "outputs": [],
   "source": [
    "cross_subgenome_order = assembl_stats.groupby(['co','subg']).apply(lambda x: x.iloc[0])['cross_subg']\n",
    "\n",
    "fig, axes = plt.subplots(ncols=3, figsize=[13,4])\n",
    "\n",
    "for m, ax in zip(['size_ratio','N50','n'], axes):\n",
    "\n",
    "    #sns.violinplot(x='cross_subg', y=m, data=assembl_stats, color='0.9',\n",
    "    #            scale='width', order=cross_subgenome_order, ax=ax)\n",
    "    sns.stripplot(x='cross_subg', y=m, data=assembl_stats, color='k', alpha=0.3,\n",
    "                order=cross_subgenome_order, ax=ax)\n",
    "    ax.set_xticklabels(cross_subgenome_order, rotation=90, size=8)\n",
    "    \n",
    "    ax.axhline(assembl_stats[m].mean(), ls='--', c='red', lw=1, zorder=10)\n",
    "\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.savefig('/mnt/HDD3/lrma/fig/draft_assembly_stats.png', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d43b8fb-4c69-4fce-993c-598fa0ffcc2c",
   "metadata": {},
   "source": [
    "# analysis of structural variants from draft assemblies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d07ec93-e2f5-40c8-b319-707b46b73e6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "NEW_ASSEMBLIES = {}\n",
    "tig_order = {}\n",
    "cum_len = {}\n",
    "for s in set(ns_subg['subg'].values):\n",
    "    NEW_ASSEMBLIES[s] = {}\n",
    "    tig_order[s] = []\n",
    "    cl = []\n",
    "    path = f'/home/mathieu/paradoxus_nanopore/MA_parents/assemblies/{s}.chromosomes.fasta'\n",
    "    \n",
    "    with open(path) as fi:\n",
    "        for seq in SeqIO.parse(fi, 'fasta'):\n",
    "            NEW_ASSEMBLIES[s][seq.id] = seq\n",
    "            tig_order[s].append(seq.id)\n",
    "            cl.append(len(seq.seq))\n",
    "    \n",
    "    cl = pd.Series([0]+list(np.cumsum(cl)[:-1]), index=tig_order[s])\n",
    "    cum_len[s] = cl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c94fbd53-28bb-456c-b02b-78e44826a459",
   "metadata": {},
   "outputs": [],
   "source": [
    "SV_draft = []\n",
    "for s in ns_subg.index:\n",
    "    df = pd.read_csv(f'/mnt/HDD3/lrma/svim-asm_draft/{s}/variants.vcf', skiprows=34, sep='\\t')\n",
    "    if df.shape[0] > 0:\n",
    "        df['strain'] = s\n",
    "        SV_draft.append(df)\n",
    "SV_draft = pd.concat(SV_draft).reset_index(drop=True)\n",
    "SV_draft = pd.concat([SV_draft, SV_draft['INFO'].apply(lambda x: pd.Series(dict([i.split('=') for i in x.split(';')])))], axis=1)\n",
    "SV_draft = SV_draft.astype({'SVLEN':np.float32})\n",
    "col1 = list(assembl_stats.columns)\n",
    "col2 = [i for i in ns_subg if i not in col1]\n",
    "for s, df in SV_draft.groupby('strain'):\n",
    "    if s in assembl_stats.index:\n",
    "        for c in col1:\n",
    "            SV_draft.loc[df.index, c] = assembl_stats.loc[s, c]\n",
    "    for c in col2:\n",
    "        SV_draft.loc[df.index, c] = ns_subg.loc[s, c]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d675fe93-7a3a-4645-ba21-cdd2f55f4c49",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add absolute coordinates to SVs\n",
    "for (subg,chrom), df in SV_draft.groupby(['subg','#CHROM']):\n",
    "    SV_draft.loc[df.index, 'pos'] = df['POS'] + cum_len[subg][chrom]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f086b60a-c792-42b3-9481-91cb11c9836a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#classify breakpoints as private\n",
    "SV_draft.loc[SV_draft['identity_filter'], 'min_dist'] = 12e6\n",
    "for (svt,subg,chrom), df in SV_draft.loc[SV_draft['identity_filter']].groupby(['SVTYPE','subg','#CHROM']):\n",
    "    \n",
    "    for i,j in itertools.combinations(df.index, 2):\n",
    "        #ensure that comparison is between different strains\n",
    "        if df.loc[i,'s_subg'] != df.loc[j,'s_subg']:\n",
    "            d = abs(df.loc[i,'POS'] - df.loc[j,'POS'])\n",
    "            if d < SV_draft.loc[i, 'min_dist']:\n",
    "                SV_draft.loc[i, 'min_dist'] = d\n",
    "            if d < SV_draft.loc[j, 'min_dist']:\n",
    "                SV_draft.loc[j, 'min_dist'] = d\n",
    "SV_draft['log_min_dist'] = np.log10(SV_draft['min_dist']+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9067860a-91aa-4c8d-8a31-0aeb12b1bcfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(x='log_min_dist', hue='subg', bins=100, data=SV_draft.loc[SV_draft['SVTYPE']=='BND'])\n",
    "\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bcf6b4f-ce85-4273-9336-d04306f991ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "hist_bp = {}\n",
    "bins = np.arange(0,12e6,1e4)\n",
    "for subg, df in SV_draft.loc[SV_draft['SVTYPE']=='BND'].groupby('subg'):\n",
    "    hist_bp[subg] = []\n",
    "    for s, df1 in df.groupby('s_subg'):\n",
    "        h = np.histogram(df1['pos'], bins=bins)\n",
    "        h = pd.Series(h[0], name=s)\n",
    "        hist_bp[subg].append(h)\n",
    "    hist_bp[subg] = pd.concat(hist_bp[subg], axis=1).T\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32a29d03-7386-491e-afb9-2b28cc50a3d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subg, df in hist_bp.items():\n",
    "    fig, ax = plt.subplots(figsize=[10,3])\n",
    "    sns.heatmap(df, cmap='Blues', ax=ax)\n",
    "    ax.set_title(subg)\n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65dd09f2-9c4d-4610-b447-c6cddd9c7239",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(x='SVLEN', hue='subg', data=SV_draft.loc[SV_draft['SVTYPE']=='INS'], bins=100)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
