{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdd7a59c-700c-49da-88bc-afaa2d1f326e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import pysam\n",
    "import pickle as pkl\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.patches import FancyArrow\n",
    "from matplotlib.patches import Rectangle\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib import cm\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.colors import to_rgba\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO\n",
    "from Bio import SeqRecord\n",
    "from Bio import Seq\n",
    "from scipy import stats\n",
    "from statsmodels.api import formula as smf\n",
    "from progressbar import ProgressBar\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "from roman import toRoman\n",
    "import itertools\n",
    "import intervaltree\n",
    "import networkx as nx\n",
    "from collections import Counter\n",
    "import re\n",
    "import random\n",
    "import gzip\n",
    "rc_sns = {'ytick_color':'k', 'xtick_color':'k', 'text_color':'k', 'font.sans-serif':'DejaVu Sans', 'figure.facecolor':(1,1,1,1)}\n",
    "sns.set_style(style='ticks', rc=rc_sns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cef8cd26-ab30-4a6b-bc42-eab493f99914",
   "metadata": {},
   "source": [
    "# Import metadata and define general variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc7e689f-31bb-4551-a404-9526a58644b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#path for figures\n",
    "fig_path = '/home/mathieu/mhenault_landrylab/Publications/lrma/draft/fig/'\n",
    "tables_path = '/home/mathieu/mhenault_landrylab/Publications/lrma/draft/tables/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99ebd6fc-aa9a-4cb6-b01f-202f2027e74d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#import tables of strain identities and cross parents\n",
    "nano_strains = pd.read_csv('/mnt/HDD3/lrma/script/nano_strains.csv', index_col=0)\n",
    "Cross_parents = pd.read_csv('/mnt/HDD3/lrma/private_variants/cross_parents.txt', sep=';', header=None, index_col=0).squeeze()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22965866-ff96-4e86-95bb-cb5b483cdd46",
   "metadata": {},
   "outputs": [],
   "source": [
    "cross_alias = {'VL3':'CC1',\n",
    "              'VL4':'CC2',\n",
    "              'VL5':'CC3',\n",
    "              'VL1':'BB1',\n",
    "              'VL2':'BB2',\n",
    "              'L1':'BC1',\n",
    "              'L2':'BC2',\n",
    "              'M1':'BA1',\n",
    "              'M2':'BA2',\n",
    "              'H1':'BSc1',\n",
    "              'H2':'BSc2'}\n",
    "nano_strains['cross'] = nano_strains['cross'].replace(cross_alias)\n",
    "cross_order = {j:i for i,j in enumerate(['CC1', 'CC2', 'CC3', 'BB1', 'BB2', 'BC1', 'BC2', 'BA1', 'BA2', 'BSc1', 'BSc2'])}\n",
    "cross_color = nano_strains.groupby('cross')['cross_color'].apply(lambda x: x.iloc[0]).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d40601fd-3eca-441d-9b2c-a10e66a7e99a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import tables of strain identities and cross parents\n",
    "nano_strains = pd.read_csv('/mnt/HDD3/lrma/script/nano_strains.csv', index_col=0)\n",
    "nano_strains['cross'] = nano_strains['cross'].replace(cross_alias)\n",
    "cross_parents = pd.read_csv('/mnt/HDD3/lrma/script/cross_parents.csv', sep=',', index_col=0)\n",
    "cross_color = nano_strains.groupby('cross')['cross_color'].apply(lambda x: x.iloc[0]).to_dict()\n",
    "\n",
    "parents_background = cross_parents.value_counts(['strain', 'group', 'mat', 'ho', 'ade2']).rename('count').reset_index().set_index('strain')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cafb5168-dec3-416e-9b2f-9a37d2d9e6e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dict of parental strains with new cross IDs\n",
    "for (cross, p) in Cross_parents.items():\n",
    "    Cross_parents[cross_alias[cross]] = p\n",
    "Parents_order = dict(zip(['LL2011_004', 'LL2011_009', 'MSH-587-1', 'LL2011_012', 'LL2011_001',\n",
    "                          'MSH-604', 'UWOPS-91-202', 'LL2012_028', 'LL2012_021',\n",
    "                          'YPS644', 'YPS744', 'LL2013_040', 'LL2013_054'], range(13)))\n",
    "\n",
    "parents_group = {'MSH-604':'SpB',\n",
    "                   'UWOPS-91-202':'SpB',\n",
    "                   'LL2012_021':'SpB',\n",
    "                   'LL2012_028':'SpB',\n",
    "                   'LL2011_004':'SpC',\n",
    "                   'LL2011_009':'SpC',\n",
    "                   'MSH-587-1':'SpC',\n",
    "                   'LL2011_012':'SpC',\n",
    "                   'LL2011_001':'SpC',\n",
    "                   'YPS644':'SpA',\n",
    "                   'YPS744':'SpA',\n",
    "                   'LL2013_040':'Sc',\n",
    "                   'LL2013_054':'Sc'}\n",
    "\n",
    "parents_color = {'MSH-604':'red',\n",
    "                   'UWOPS-91-202':'red',\n",
    "                   'LL2012_021':'darkred',\n",
    "                   'LL2012_028':'darkred',\n",
    "                   'LL2011_004':'dodgerblue',\n",
    "                   'LL2011_009':'dodgerblue',\n",
    "                   'MSH-587-1':'midnightblue',\n",
    "                   'LL2011_012':'midnightblue',\n",
    "                   'LL2011_001':'midnightblue',\n",
    "                   'YPS644':'limegreen',\n",
    "                   'YPS744':'limegreen',\n",
    "                   'LL2013_040':'dimgrey',\n",
    "                   'LL2013_054':'dimgrey'}\n",
    "\n",
    "mat_alias = {'a': r'$a$', 'alpha':r'$\\alpha$'}\n",
    "group_alias = {'SpA':r'$SpA$', 'SpB':r'$SpB$', 'SpC':r'$SpC$', 'Scer':r'$S.c.$'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9e5ff8a-e74e-4c37-a6ef-1a6787bdabb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# list parents\n",
    "Parents = list(nano_strains.loc[nano_strains['cross']=='P', 'strain'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "011df572-3bba-473e-8b4d-ff80fe8438d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ty_order = ['Ty1p', 'Ty3p', 'Tsu4', 'TY1', 'TY2', 'TY3', 'TY4']\n",
    "\n",
    "ty_colors = {'Ty1p':'blue',\n",
    "             'Ty3p':'forestgreen',\n",
    "             'Tsu4':'red',\n",
    "             'Ty5p':'gold',\n",
    "             'TY1':'dodgerblue',\n",
    "             'TY2':'cyan',\n",
    "             'TY3':'limegreen',\n",
    "             'TY4':'lightcoral',\n",
    "             'TY5':'yellow'}\n",
    "\n",
    "ty_alias = {'Ty1p':r'Ty1 Sp',\n",
    "             'Ty3p':'Ty3 Sp',\n",
    "             'Tsu4':'Tsu4',\n",
    "             'Ty5p':'Ty5 Sp',\n",
    "             'TY1':'Ty1 Sc',\n",
    "             'TY2':'Ty2',\n",
    "             'TY3':'Ty3 Sc',\n",
    "             'TY4':'Ty4',\n",
    "             'TY5':'Ty5 Sc'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7094b909-25c1-4ec4-9538-771dba1fa7ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# color dictionary for annot types\n",
    "at_color = {'fl':'red', 'solo':'blue', 'tr':'green'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33eb82ef-f669-43d5-8cb7-de7630935689",
   "metadata": {},
   "outputs": [],
   "source": [
    "ns_subg = pd.read_csv('/mnt/HDD3/lrma/script/ns_subg.csv', index_col=0)\n",
    "ns_subg['cross'] = ns_subg['cross'].replace(cross_alias)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23764d20-919c-4140-9c33-ddd9183dfa04",
   "metadata": {},
   "source": [
    "# General functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b1e4239-df7b-4bac-bcf6-901086169509",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_pval_symbol(p):\n",
    "    if p<=0.001:\n",
    "        pdisplay = '***'\n",
    "    elif p<=0.01:\n",
    "        pdisplay = '**'\n",
    "    elif p<=0.05:\n",
    "        pdisplay = '*'\n",
    "    else:\n",
    "        pdisplay = None\n",
    "    return pdisplay\n",
    "\n",
    "def plot_pval_text(p):\n",
    "    if p<=0.001:\n",
    "        pdisplay = f'{p:.1e}'\n",
    "    elif p<=0.01:\n",
    "        pdisplay = f'{p:.4f}'\n",
    "    elif p<=0.1:\n",
    "        pdisplay = f'{p:.3f}'\n",
    "    elif p<=1:\n",
    "        pdisplay = f'{p:.2f}'\n",
    "    elif p>1:\n",
    "        pdisplay = f'{p:.1f}'\n",
    "    else:\n",
    "        pdisplay = None\n",
    "    return pdisplay"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb080d51-5465-49b4-a8ed-296e92e64b9d",
   "metadata": {},
   "source": [
    "# Import chromosomes and contig offsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af040496-fa9c-42ef-b93a-7a69c58dbdc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "tig_off = {}\n",
    "for r in ns_subg['subg'].unique():\n",
    "    path = f'/home/mathieu/paradoxus_nanopore/MA_parents/assemblies/{r}.chromosomes.rdna.fasta'\n",
    "    \n",
    "    df = pd.concat([pd.Series([seq.id, len(seq.seq)]) for seq in SeqIO.parse(path, 'fasta')], axis=1).T\n",
    "    df.index = df[0].values\n",
    "    df[2] = np.concatenate([np.array([0]), np.cumsum(df[1].values)[:-1]])\n",
    "    df[3] = df[1] + df[2]\n",
    "    df['color'] = np.tile([0,1], 10)[:df.shape[0]]\n",
    "    df['rank'] = range(df.shape[0])\n",
    "    \n",
    "    tig_off[r] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60d3d8ba-762f-4e93-9c4b-b44dbcffa4ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tract definition done in a separate script\n",
    "# MA lines subgenomes + parents\n",
    "\n",
    "Tracts = []\n",
    "\n",
    "# MA lines subgenomes\n",
    "for s in ns_subg.loc[ns_subg['identity_filter']==True].index:\n",
    "    tracts = pd.read_csv(f'/mnt/HDD3/lrma/depth_tracts/{s}.tracts.csv', index_col=0)\n",
    "    subg = s.split('.')[1]\n",
    "    tracts['subg'] = subg\n",
    "    \n",
    "    for chrom, df in tracts.groupby('chrom'):\n",
    "        tracts.loc[df.index, 'Start'] = df['start'] + tig_off[subg].loc[chrom, 2]\n",
    "        tracts.loc[df.index, 'End'] = df['end'] + tig_off[subg].loc[chrom, 2]\n",
    "    \n",
    "    Tracts.append(tracts)\n",
    "\n",
    "# parents\n",
    "for s in nano_strains.loc[nano_strains['cross']=='P'].index:\n",
    "    tracts = pd.read_csv(f'/mnt/HDD3/lrma/depth_tracts/{s}.tracts.csv', index_col=0)\n",
    "    subg = s\n",
    "    tracts['subg'] = subg\n",
    "    \n",
    "    for chrom, df in tracts.groupby('chrom'):\n",
    "        tracts.loc[df.index, 'Start'] = df['start'] + tig_off[subg].loc[chrom, 2]\n",
    "        tracts.loc[df.index, 'End'] = df['end'] + tig_off[subg].loc[chrom, 2]\n",
    "        \n",
    "    Tracts.append(tracts)\n",
    "        \n",
    "Tracts = pd.concat(Tracts).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4ed48ef-5b53-4c71-bf66-e9257ab11f2a",
   "metadata": {},
   "source": [
    "# Import genome annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ff154a7-886a-4be0-8898-d7be118e3de1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "GFF = {}\n",
    "nano_strains_annot = pd.read_csv('/mnt/HDD1/paradoxus_nanopore/MA_parents/annotation/nanopore_assemblies.csv', index_col=0)\n",
    "# create a dictionary for later conversion of assembly names\n",
    "assembly_background_conversion = nano_strains_annot.loc[nano_strains_annot['ploidy']=='haploid', 'background']\n",
    "assembly_background_conversion = {j:i for (i,j) in assembly_background_conversion.items()}\n",
    "\n",
    "for s in nano_strains_annot.index:\n",
    "\n",
    "    gff_path = f'/home/mathieu/paradoxus_nanopore/MA_parents/annotation/combined_gff/{s}.gff'\n",
    "    if '.' in s:\n",
    "        tig_begin = 'chr'\n",
    "    else:\n",
    "        tig_begin = 'utg'\n",
    "    \n",
    "    with open(gff_path) as handle:\n",
    "        gff = pd.DataFrame([line.split('\\t') for line in handle.read().splitlines() if line[:3]==tig_begin])\n",
    "        gff = gff.loc[gff[1]=='maker']\n",
    "        gff = gff.astype({3:int, 4:int})\n",
    "        GFF[s] = gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "451e0a02-d103-4c92-84e4-94fd465741f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import REA\n",
    "REA = pd.read_csv('/mnt/HDD3/lrma/results/REA.csv', index_col=0)\n",
    "# add single filter for truncated elements\n",
    "REA['tr'] = (REA['fl_tr']==True) & (REA['fl']==False)\n",
    "rea_reindex = REA.loc[~REA['og'].isna()].set_index(['og','s_subg'], drop=False).sort_index(level=['og','s_subg'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96acc016-8811-49c1-ae2c-62963d6f56d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "lines_per_cross = ns_subg.loc[ns_subg['identity_filter']==True].groupby('cross').apply(lambda x: set(x['strain'].values)).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdcd9dd6-f358-45ff-b2c2-b49c830ca884",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_data = pd.read_csv('/mnt/HDD3/lrma/results/og_data.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecd38582-eb71-46dd-bf5f-86ab655052be",
   "metadata": {},
   "source": [
    "## Visualization of LOH & aneuploidies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d936030c-6c52-4c78-958a-645d45e34577",
   "metadata": {},
   "outputs": [],
   "source": [
    "cn_color = dict(zip(range(5), [cm.coolwarm(i) for i in np.concatenate([np.array([0]), np.linspace(0.5,1,4)])]))\n",
    "\n",
    "def plot_subg_cn(tracts, y, s, w=0.2):\n",
    "    \n",
    "    max_cn = len(cn_color)-1\n",
    "    \n",
    "    for i in tracts.index:\n",
    "        Start, End, call = tracts.loc[i, ['Start','End','call']]\n",
    "        if call > max_cn:\n",
    "            call = max_cn\n",
    "        c = cn_color[call]\n",
    "\n",
    "        fa = FancyArrow(Start, y, End-Start, 0, fc=c, width=w, head_width=0, head_length=0, lw=0)\n",
    "        ax.add_patch(fa)\n",
    "\n",
    "    ax.text(-1.5e4, y, s, ha='right', va='center', size=7)\n",
    "\n",
    "def plot_chromosomes(s, y, w, plot_text=True):\n",
    "    color_dict={0:'0.4', 1:'0'}\n",
    "    to = tig_off[s]\n",
    "    for chrom in to.index:\n",
    "        start, end, c = to.loc[chrom, [2,3,'color']]\n",
    "        c = color_dict[c]\n",
    "        fa = FancyArrow(start, y, end-start, 0, fc=c, width=w, head_width=0, head_length=0, lw=0)\n",
    "        ax.add_patch(fa)\n",
    "        if plot_text==True:\n",
    "            ax.text(np.mean([start, end]), y, chrom.strip('chr'), color='white', ha='center', va='center', size=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4654f790-e609-40bb-bab2-49097b17ef88",
   "metadata": {},
   "outputs": [],
   "source": [
    "for cross, df in nano_strains.loc[(nano_strains['identity_filter']==True) &\n",
    "                                 (nano_strains['cross'].isin(cross_order))].groupby('cross'):\n",
    "\n",
    "    parents = Cross_parents[cross].split(',')\n",
    "    lines_order = parents + list(df['strain'].apply(lambda x: int(x[1:])).sort_values().index)\n",
    "    lines_order = {j:i for i,j in enumerate(lines_order)}\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=[7,6], \n",
    "                           gridspec_kw=dict(left=0.15, right=0.9, top=0.96, bottom=0.08))\n",
    "\n",
    "    for s, Y in lines_order.items():\n",
    "\n",
    "        if s in parents:\n",
    "            y = Y\n",
    "            tracts = Tracts.loc[Tracts['s_subg']==s]\n",
    "\n",
    "            plot_subg_cn(tracts, y+0.22, s, w=0.4)\n",
    "\n",
    "            plot_chromosomes(s, y-0.22, 0.4)\n",
    "\n",
    "\n",
    "\n",
    "        else:\n",
    "\n",
    "            for p, dy in zip(parents, [-0.22, 0.22]):\n",
    "                y = Y+dy\n",
    "                s_subg = f'{s}.{p}'\n",
    "                tracts = Tracts.loc[Tracts['s_subg']==s_subg]\n",
    "\n",
    "                plot_subg_cn(tracts, y+0.1, s_subg)\n",
    "                plot_chromosomes(p, y-0.05, 0.1, plot_text=False)\n",
    "\n",
    "    ax.set_ylim(-0.5, len(lines_order)-0.5)\n",
    "    ax.set_yticks([])\n",
    "    ax.set_xlim(-1e4,12e6)\n",
    "    ax.set_title(cross)\n",
    "\n",
    "    #add legend\n",
    "\n",
    "    legend_elms = [Line2D([0], [0], color='w', marker='s', mfc=c, label=l) for (l,c) in cn_color.items()]\n",
    "\n",
    "    ax.legend(handles=legend_elms, bbox_to_anchor=[0.98, 0.5], loc=6)\n",
    "\n",
    "    #plt.savefig(f'/mnt/HDD1/Dropbox/fig_lrma/cn_tracts_{cross}.png', dpi=300)\n",
    "    #plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a80fa0c7-6e89-41b8-9ad1-b3c9096e717e",
   "metadata": {},
   "outputs": [],
   "source": [
    "tracts_trees = {s_subg:{} for s_subg in Tracts['s_subg'].unique()}\n",
    "for (s, chrom), df in Tracts.groupby(['s_subg', 'chrom']):\n",
    "    t = intervaltree.IntervalTree()\n",
    "    for i in df.index:\n",
    "        start, end, call = df.loc[i, ['start','end','call']].values\n",
    "        t[start:end] = (i, call)\n",
    "    tracts_trees[s][chrom] = t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db78d31e-493c-4a9e-8687-9210e406c9fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "tracts_trees['E89.YPS744']['chrI']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94dd59e4-040d-4c29-a6f2-1b660b6fe7a6",
   "metadata": {},
   "source": [
    "# Define orthology between annots by cross\n",
    "\n",
    "## Dot plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef08209c-4384-4b95-8fac-2bb814eb768e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for cross in cross_order: \n",
    "    \n",
    "    p1, p2 = Cross_parents.loc[cross].split(',')\n",
    "\n",
    "    coords = pd.read_csv(f'/mnt/HDD3/lrma/mummer_crosses/{cross}.coords', sep='\\t', header=None)\n",
    "    coords.columns = ['ref_start', 'ref_end', 'qry_start', 'qry_end', 'ref_len', 'qry_len', 'ident',\n",
    "                      'ref_strand', 'qry_strand', 'ref_name', 'qry_name']\n",
    "\n",
    "    coords['Ref_start'] = tig_off[p1].loc[coords['ref_name'], 2].values + coords['ref_start'].values\n",
    "    coords['Ref_end'] = tig_off[p1].loc[coords['ref_name'], 2].values + coords['ref_end'].values\n",
    "    coords['Qry_start'] = tig_off[p2].loc[coords['qry_name'], 2].values + coords['qry_start'].values\n",
    "    coords['Qry_end'] = tig_off[p2].loc[coords['qry_name'], 2].values + coords['qry_end'].values\n",
    "    #only keep chromosome-to-chromosome alignments\n",
    "    coords = coords.loc[coords['ref_name']==coords['qry_name']]\n",
    "\n",
    "    toq = tig_off[p2]\n",
    "    tor = tig_off[p1]\n",
    "\n",
    "    for tig, df in coords.groupby('ref_name'):\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=[4,4], gridspec_kw={'left':0.15, 'right':0.85, 'top':0.85, 'bottom':0.15})\n",
    "\n",
    "        for i in df.index:\n",
    "            rs, Re, qs, qe = df.loc[i, ['Ref_start', 'Ref_end', 'Qry_start', 'Qry_end']]\n",
    "\n",
    "\n",
    "            ax.plot([rs, Re], [qs, qe], c='k', lw=0.5, zorder=2)\n",
    "\n",
    "        # plot annots for each parent\n",
    "        tor_min, tor_max = tor.loc[tig, [2,3]]\n",
    "        toq_min, toq_max = toq.loc[tig, [2,3]]\n",
    "\n",
    "        rea = REA.loc[(REA['s_subg']==p1) & (REA['fl']) & (REA['Query']==tig)]\n",
    "        if rea.shape[0] > 0:\n",
    "            for i in rea.index:    \n",
    "                fam, Chrom, Start, End, Strand, og = rea.loc[i, ['family', 'Query', 'Start', 'End', 'Strand', 'og']]\n",
    "                c = ty_colors[fam]\n",
    "                x = np.mean([Start,End])+tor.loc[Chrom, 2]\n",
    "                ax.axvline(x, zorder=3, color=c, lw=0.5)\n",
    "                ax.text(x, toq_max, f'{Strand} {og}', ha='left', va='bottom', size=6, color=c, rotation=90)\n",
    "\n",
    "        rea = REA.loc[(REA['s_subg']==p2) & (REA['fl']) & (REA['Query']==tig)]\n",
    "        if rea.shape[0] > 0:\n",
    "            for i in rea.index:    \n",
    "                fam, Chrom, Start, End, Strand, og = rea.loc[i, ['family', 'Query', 'Start', 'End', 'Strand', 'og']]\n",
    "                c = ty_colors[fam]\n",
    "                y = np.mean([Start,End])+toq.loc[Chrom, 2]\n",
    "                ax.axhline(y, zorder=3, color=ty_colors[fam], lw=0.5)\n",
    "                ax.text(tor_max, y, f'{Strand} {og}', ha='left', va='bottom', size=6, color=c)\n",
    "\n",
    "        \n",
    "        ax.set_xlim(tor_min, tor_max)\n",
    "        xticks = np.arange(tor_min, tor_max, 1e5)\n",
    "        ax.set_xticks(xticks)\n",
    "        ax.set_xticklabels(np.int16((xticks-tor_min)*1e-5))\n",
    "        ax.set_xlabel(f'{p1} (100 kb)')\n",
    "\n",
    "        ax.set_ylim(toq_min, toq_max)\n",
    "        yticks = np.arange(toq_min, toq_max, 1e5)\n",
    "        ax.set_yticks(yticks)\n",
    "        ax.set_yticklabels(np.int16((yticks-toq_min)*1e-5))\n",
    "        ax.set_ylabel(f'{p2} (100 kb)')\n",
    "\n",
    "        ax.set_title(f'{cross} {tig}')\n",
    "\n",
    "        #plt.show()\n",
    "        #plt.savefig(f'/home/mathieu/mhenault_landrylab/Publications/lrma/pre-draft/fig/orthology_parents/{cross}_{tig}.png', dpi=300)\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be0fcd5f-17b8-4912-9de2-2d1c6e188902",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import orthologous parental loci\n",
    "parental_orthology = pd.read_csv('/mnt/HDD3/lrma/results/parental_orthology.csv')\n",
    "parental_orthology.set_index('cross', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "789b95f1-ac04-4fc3-8ee3-67f983e263b2",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Genome annotation summary\n",
    "## Fig S4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8661fa36-aa85-44d9-9be3-66650f756b89",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=[9,9])\n",
    "gs = plt.GridSpec(ncols=3, nrows=2, height_ratios=[7,3], width_ratios=[1,3,1],\n",
    "                 hspace=0.18, top=0.95, bottom=0.05, right=0.85, left=0.07)\n",
    "\n",
    "# Plot annotations\n",
    "ax = fig.add_subplot(gs[0, :])\n",
    "\n",
    "chrom_color = {0:'0.75', 1:'0.9'}\n",
    "strand_dy = {'+':-0.15, '-':+0.15}\n",
    "strand_marker = {'+': '>', '-': '<'}\n",
    "    \n",
    "for (cross, Y) in cross_order.items():\n",
    "    Y*=2\n",
    "    for y, p in zip([Y-0.35, Y+0.35],\n",
    "                    Cross_parents[cross].split(',')):\n",
    "    \n",
    "        rea = REA.loc[(REA['s_subg']==p) & (REA['fl'])]\n",
    "\n",
    "        if rea.shape[0] > 0:\n",
    "\n",
    "            for i in rea.index:    \n",
    "                fam, Chrom, Start, End, Strand = rea.loc[i, ['family', 'Query', 'Start', 'End', 'Strand']]\n",
    "                ax.scatter(np.mean([Start,End])+tig_off[p].loc[Chrom, 2], y+strand_dy[Strand], zorder=3, \n",
    "                            marker=strand_marker[Strand], color=ty_colors[fam], lw=0, s=30, alpha=0.7)\n",
    "\n",
    "        for (chrom, length, start, color), df in tig_off[p].groupby([0,1,2,'color']):\n",
    "\n",
    "            ar = FancyArrow(start, y, length, 0, width=0.5, head_width=1, zorder=0,\n",
    "                            fc=chrom_color[color], lw=0, length_includes_head=True, clip_on=False, head_length=0)\n",
    "            ax.add_patch(ar)\n",
    "            ax.text(start+0.5*length, y, chrom[3:], ha='center', va='center', size=6, color='k', zorder=1)\n",
    "        \n",
    "        # add strain name label to the right\n",
    "        l = f'{p} ({group_alias[parents_background.loc[p, \"group\"]]})'\n",
    "        ax.text(tig_off[p][3].max()+5e4, y, l, size=8, ha='left', va='center')\n",
    "            \n",
    "    # add annotation of orthology\n",
    "    \n",
    "    if cross in parental_orthology.index:\n",
    "        df = parental_orthology.loc[[cross]].reset_index()\n",
    "        p1, p2 = Cross_parents.loc[cross].split(',')\n",
    "        for i in df.index:\n",
    "            og_p1, og_p2 = df.loc[i, ['og_p1', 'og_p2']]\n",
    "            chrom_p1, pos_p1, fam = og_data.loc[og_p1, ['Query','og_mid','family']]\n",
    "            chrom_p2, pos_p2 = og_data.loc[og_p2, ['Query','og_mid']]\n",
    "            pos_p1 += tig_off[p1].loc[chrom_p1, 2]\n",
    "            pos_p2 += tig_off[p2].loc[chrom_p2, 2]\n",
    "            \n",
    "            c = ty_colors[fam]\n",
    "            hw = 6e4\n",
    "            x1, x2, x3, x4 = (pos_p1-hw, pos_p1+hw, pos_p2-hw, pos_p2+hw)\n",
    "            y1, y2, y3, y4 = (-0.65, -0.1, 0.1, 0.65)\n",
    "            xy = np.stack([np.array([x1,x2,x2,x4,x4,x3,x3,x1,x1]),\n",
    "                           np.array([y1,y1,y2,y3,y4,y4,y3,y2,y1])+Y]).T\n",
    "            R = Polygon(xy, lw=0, fc=c, zorder=2, alpha=0.3)\n",
    "            ax.add_patch(R)\n",
    "            ax.text(pos_p2, 0.4+Y, '*', ha='center', va='bottom', fontweight='bold', c=c)\n",
    "            \n",
    "ax.set_ylim(-1, 21)\n",
    "ax.set_yticks(range(0,22,2))\n",
    "ax.set_yticklabels(pd.Series(cross_order).sort_values().index)\n",
    "ax.set_xlim(-1e5, 12e6)\n",
    "ax.set_xticks(np.arange(0, 12.1e6, 1e6))\n",
    "ax.set_xticklabels(range(13))\n",
    "ax.set_xlabel('Mb')\n",
    "\n",
    "#Legend of Ty families\n",
    "legend_elms = [Line2D([0], [0], color='w', marker='^', ms=10, mfc=ty_colors[ty], label=ty_alias[ty]) for ty in ty_order if '5' not in ty]\n",
    "ax.legend(handles=legend_elms, loc='lower center', bbox_to_anchor=[0.55, 1], ncols=7,\n",
    "          handlelength=1, handletextpad=0.5, columnspacing=1)\n",
    "\n",
    "sns.despine(ax=ax)\n",
    "\n",
    "# Heatmap of annotation counts\n",
    "\n",
    "ax = fig.add_subplot(gs[1, 1])\n",
    "cbar_ax = fig.add_axes([0.78, 0.08, 0.02, 0.2])\n",
    "\n",
    "rea = REA.loc[REA['cross']=='P']\n",
    "\n",
    "dat = rea.loc[rea['fl']].value_counts(['family','s_subg']).rename('count').reset_index().pivot_table(index='s_subg', columns='family', values='count').fillna(0)\n",
    "dat.loc['UWOPS-91-202'] = 0\n",
    "dat['Ty5p'] = 0\n",
    "dat['TY5'] = 0\n",
    "\n",
    "sns.heatmap(dat.loc[Parents_order, ty_order], cmap='Blues', annot=True, ax=ax, vmin=0, vmax=40,\n",
    "            annot_kws={'size':8}, cbar_ax=cbar_ax, cbar_kws={'orientation':'vertical', 'label':'Full-length Ty count'})\n",
    "\n",
    "ax.set_ylabel('')\n",
    "ax.set_yticks(np.arange(13)+0.5)\n",
    "ax.set_yticklabels([f'{p} ({group_alias[parents_background.loc[p, \"group\"]]})' for p in Parents_order], size=8)\n",
    "ax.invert_yaxis()\n",
    "\n",
    "ax.set_xticklabels([ty_alias[ty] for ty in ty_order])\n",
    "ax.set_xlabel('')\n",
    "\n",
    "cbar_ax.spines['left'].set_visible(False)\n",
    "cbar_ax.spines['bottom'].set_visible(False)\n",
    "\n",
    "fig.text(0.02, 0.96, 'A', size=24, fontweight='bold', font='Arial')\n",
    "fig.text(0.02, 0.28, 'B', size=24, fontweight='bold', font='Arial')\n",
    "\n",
    "#for ext in ['svg', 'png']:\n",
    "#    plt.savefig(f'{fig_path}Supp_FigS4.{ext}', dpi=300)        \n",
    "    \n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6984cb9-2061-4c3d-bd05-caeb08223a10",
   "metadata": {},
   "source": [
    "# Classification of orthogroups\n",
    "## Create table to describe orthogroup state for each line subgenome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86fa9bee-27a0-453a-8132-80e0587d6ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_cn = []\n",
    "\n",
    "\n",
    "for (og, subg, og_contains_parent, chrom, left_bound, right_bound), df in og_data.loc[og_data['og_fl']]\\\n",
    ".groupby(['og', 'og_subg', 'og_contains_parent', 'Query', 'og_left_bound', 'og_right_bound']):\n",
    "     \n",
    "    rea = rea_reindex.loc[og]\n",
    "    crosses = rea['cross'].unique()\n",
    "    \n",
    "    # get all lines subgenomes (plus the parental subgenome) that could have annotations\n",
    "    S = []\n",
    "    for cross in crosses:\n",
    "        if cross != 'P':\n",
    "            S.extend([f'{s}.{subg}' for s in lines_per_cross[cross]])\n",
    "    if og_contains_parent:\n",
    "        S.append(subg)\n",
    "    \n",
    "    for s in S:\n",
    "        \n",
    "        left_cn = np.nan\n",
    "        right_cn = np.nan\n",
    "        \n",
    "        left_tract = tracts_trees[s][chrom][left_bound]\n",
    "        if len(left_tract) == 1:\n",
    "            left_cn = list(left_tract)[0][2][1]\n",
    "            \n",
    "        right_tract = tracts_trees[s][chrom][right_bound]\n",
    "        if len(right_tract) == 1:\n",
    "            right_cn = list(right_tract)[0][2][1]\n",
    "            \n",
    "        if s in rea.index:\n",
    "            sub_rea= rea.loc[s]\n",
    "            has_annot = True\n",
    "            has_fl = sub_rea['fl'].sum() > 0\n",
    "        else:\n",
    "            has_annot = False\n",
    "            has_fl = False\n",
    "        \n",
    "        og_cn.append([s, og, left_cn, right_cn, has_annot, has_fl])\n",
    "        \n",
    "og_cn = pd.DataFrame(og_cn, columns=['s_subg', 'og', 'left_cn', 'right_cn', 'has_annot', 'has_fl'])\n",
    "og_cn['cons_cn'] = (og_cn['left_cn'] == og_cn['right_cn'])\n",
    "\n",
    "og_cn['cn'] = og_cn[['left_cn','right_cn']].mean(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e8e3fc4-b7f4-4a03-9635-3646eac0ba8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add assembly annotation from reverse liftover\n",
    "for s, df in og_cn.groupby('s_subg'):\n",
    "    if s not in nano_strains.loc[nano_strains['cross']=='P', 'strain']:\n",
    "        bed = pd.read_csv(f'/mnt/HDD3/lrma/reverse_liftover/{s}.flank_lift.bed', sep='\\t', header=None).set_index(3)\n",
    "\n",
    "        for i in df.index:\n",
    "            og = df.loc[i, 'og']\n",
    "            chrom, lb, rb = og_data.loc[og, ['Query', 'og_left_bound', 'og_right_bound']]\n",
    "\n",
    "            has_assembly_left = f'{chrom}_{lb-500:.0f}_{lb:.0f}' in bed.index\n",
    "            has_assembly_right = f'{chrom}_{rb:.0f}_{rb+500:.0f}' in bed.index\n",
    "\n",
    "            og_cn.loc[i, 'has_assembly_left'] = has_assembly_left\n",
    "            og_cn.loc[i, 'has_assembly_right'] = has_assembly_right\n",
    "\n",
    "og_cn['cons_as'] = (og_cn['has_assembly_left'] == og_cn['has_assembly_right'])\n",
    "\n",
    "og_cn.set_index(['og','s_subg'], inplace=True, drop=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "220788fb-fa12-48e1-9556-a46cdf199913",
   "metadata": {},
   "source": [
    "## Decision tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e6edb84-12dd-4aa7-8e8e-6ab77de4f897",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_annot_seq(seq0, seq1):\n",
    "\n",
    "    delta_seq = np.nan\n",
    "    which_short_seq = np.nan\n",
    "    short_seq = np.nan\n",
    "    long_seq = np.nan\n",
    "\n",
    "    if len(seq0)-len(seq1) == 1:\n",
    "        short_seq = seq1\n",
    "        long_seq = seq0\n",
    "        delta_seq = 1\n",
    "        which_long_seq = 0\n",
    "\n",
    "    elif len(seq1)-len(seq0) == 1:\n",
    "        short_seq = seq0\n",
    "        long_seq = seq1\n",
    "        delta_seq = 1\n",
    "        which_long_seq = 1\n",
    "        \n",
    "    elif len(seq0) == len(seq1):\n",
    "        delta_seq = 0\n",
    "        \n",
    "    else:\n",
    "        return tuple(['complex'])\n",
    "    \n",
    "    # situation where sequence lengths differ by one; looking for insertions or deletion of single annots\n",
    "    if delta_seq == 1:\n",
    "        complete_match = False\n",
    "        i = 0\n",
    "        #iterate through all positions for the gap\n",
    "        while complete_match == False and i <= len(long_seq):\n",
    "            new_short_seq = short_seq[:i] + ['-'] + short_seq[i:]\n",
    "            match_count = 0\n",
    "            for j,k in enumerate(new_short_seq):\n",
    "                if k != '-':\n",
    "                    if long_seq[j] == k:\n",
    "                        match_count += 1\n",
    "            if match_count == len(short_seq):\n",
    "                complete_match = True\n",
    "                return 'single insertion', which_long_seq, long_seq[i]\n",
    "\n",
    "            else:\n",
    "                i += 1\n",
    "        \n",
    "        if complete_match == False:\n",
    "            return tuple(['complex'])\n",
    "        \n",
    "    # situation where sequence lengths match; looking for single differences\n",
    "    if delta_seq == 0:        \n",
    "        matches = [j==k for (j,k) in zip(seq0, seq1)]\n",
    "        #all matches are true\n",
    "        if all(matches):\n",
    "            return tuple(['same content'])\n",
    "        \n",
    "        else:\n",
    "            #there is a single difference\n",
    "            if sum(matches) == len(seq0)-1:\n",
    "                for i, (j, k) in enumerate(zip(seq0, seq1)):\n",
    "                    if j != k:\n",
    "                        break\n",
    "                return 'single difference', seq0[i], seq1[i]\n",
    "            \n",
    "            else:\n",
    "                return tuple(['complex'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d771bbdb-d90a-4022-8b44-b97a4757ed2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reset\n",
    "og_cn['class'] = np.nan\n",
    "og_cn['delta_cn'] = np.nan\n",
    "\n",
    "# sort out inconsistent copy numbers and assembly presence\n",
    "og_cn.loc[(og_cn['cons_cn']==False) | (og_cn['cons_as']==False), 'class'] = 'inconsistent'\n",
    "\n",
    "#sort out assembly absence and copy number higher than 0\n",
    "og_cn.loc[(og_cn['class']!='inconsistent') & (og_cn['has_assembly_left']==False) & (og_cn['cn']>0), 'class'] = 'inconsistent'\n",
    "\n",
    "#sort out assembly presence and copy number of 0\n",
    "og_cn.loc[(og_cn['class']!='inconsistent') & (og_cn['has_assembly_left']==True) & (og_cn['cn']==0), 'class'] = 'inconsistent'\n",
    "\n",
    "#reattribute parental og classes\n",
    "og_cn.loc[og_cn['s_subg'].isin(Parents), 'class'] = 'parental'\n",
    "\n",
    "for (og, s_subg) in og_cn.loc[og_cn['class']!='inconsistent'].index:\n",
    "    \n",
    "    if s_subg not in Parents:\n",
    "        subg = s_subg.split('.')[1]\n",
    "        \n",
    "        og_complex, og_contains_parent, og_fl_counts = og_data.loc[og, ['og_complex', 'og_contains_parent', 'og_fl_counts']]\n",
    "        has_annot, has_assembly, cn = og_cn.loc[(og, s_subg), ['has_annot', 'has_assembly_left', 'cn']]\n",
    "        # set parental copy number to 0; to be updated later if applies\n",
    "        p_cn = 0\n",
    "        # reset parental values for safety\n",
    "        p_f1, p_solo, p_fam, p_has_fl = np.repeat(np.nan, 4)\n",
    "        \n",
    "        # Complex orthogroups\n",
    "        if og_complex:\n",
    "            \n",
    "            if og_contains_parent:\n",
    "                # get data on parental annotation\n",
    "                p_annot = [tuple(x) for x in rea_reindex.loc[(og, subg)].sort_values(by=['Start','End'])[['fl', 'solo', 'tr']].values]\n",
    "                p_has_fl = rea_reindex.loc[(og, subg), 'fl'].any()\n",
    "                p_cn = og_cn.loc[(og, subg), 'cn']\n",
    "            \n",
    "                if has_annot:\n",
    "                    s_annot = [tuple(x) for x in rea_reindex.loc[(og, s_subg)].sort_values(by=['Start','End'])[['fl', 'solo', 'tr']].values]\n",
    "                    \n",
    "                    if has_assembly:\n",
    "                        alignment = align_annot_seq(p_annot, s_annot)\n",
    "                        \n",
    "                        if alignment[0] == 'same content':\n",
    "                            if cn == p_cn:\n",
    "                                og_cn.loc[(og, s_subg), 'class'] = 'no_change'\n",
    "                            else:\n",
    "                                if p_has_fl:\n",
    "                                    og_cn.loc[(og, s_subg), 'class'] = 'aneup/LOH'\n",
    "                                else:\n",
    "                                    og_cn.loc[(og, s_subg), 'class'] = 'no_change'\n",
    "                                \n",
    "                        elif alignment[0] == 'complex':\n",
    "                            og_cn.loc[(og, s_subg), 'class'] = 'complex'\n",
    "                            \n",
    "                        elif alignment[0] == 'single difference':\n",
    "                            # if parent fl is true and s_subg solo is true, excision\n",
    "                            (p_fl, p_solo, p_tr), (s_fl, s_solo, s_tr) = alignment[1:]\n",
    "                            if all([p_fl, s_solo]):\n",
    "                                cn = 0\n",
    "                                og_cn.loc[(og, s_subg), 'class'] = 'excision'\n",
    "                            \n",
    "                            # if parent solo is true and s_subg fl is true, de novo\n",
    "                            elif all([p_solo, s_fl]):\n",
    "                                p_cn = 0\n",
    "                                og_cn.loc[(og, s_subg), 'class'] = 'de_novo'\n",
    "                                \n",
    "                            elif all([p_fl, s_tr]):\n",
    "                                cn = cn/2\n",
    "                                og_cn.loc[(og, s_subg), 'class'] = 'truncation'\n",
    "                                \n",
    "                            else:\n",
    "                                og_cn.loc[(og, s_subg), 'class'] = 'inconsistent'\n",
    "                                \n",
    "                        elif alignment[0] == 'single insertion':\n",
    "                            which_insert, (fl, solo, tr) = alignment[1:]\n",
    "                            # insertion in parent\n",
    "                            if which_insert == 0:\n",
    "                                # all the insertions come from here\n",
    "                                if fl:\n",
    "                                    cn = 0\n",
    "                                    og_cn.loc[(og, s_subg), 'class'] = 'deletion'\n",
    "                                elif solo:\n",
    "                                    og_cn.loc[(og, s_subg), 'class'] = 'solo_deletion'\n",
    "                                else:\n",
    "                                    og_cn.loc[(og, s_subg), 'class'] = 'inconsistent'\n",
    "                            # insertion in s_subg\n",
    "                            elif which_insert == 1:\n",
    "                                if fl:\n",
    "                                    p_cn = 0\n",
    "                                    og_cn.loc[(og, s_subg), 'class'] = 'de_novo'\n",
    "                                elif solo:\n",
    "                                    og_cn.loc[(og, s_subg), 'class'] = 'solo_insertion'\n",
    "                                else:\n",
    "                                    og_cn.loc[(og, s_subg), 'class'] = 'inconsistent'\n",
    "                        \n",
    "                    else:\n",
    "                        og_cn.loc[(og, s_subg), 'class'] = 'inconsistent'\n",
    "\n",
    "                else:\n",
    "                    cn = 0\n",
    "                    if has_assembly:\n",
    "                        og_cn.loc[(og, s_subg), 'class'] = 'deletion'\n",
    "                        \n",
    "                    else:\n",
    "                        if p_has_fl:\n",
    "                            og_cn.loc[(og, s_subg), 'class'] = 'aneup/LOH'\n",
    "                        else:\n",
    "                            og_cn.loc[(og, s_subg), 'class'] = 'no_change'\n",
    "        \n",
    "        # Non-complex orthogroups\n",
    "        else:\n",
    "            # extract annotation details for non-complex orthogroup\n",
    "            if has_annot:\n",
    "                s_fl, s_solo, s_tr = rea_reindex.loc[(og, s_subg), ['fl', 'solo', 'tr']].iloc[0]\n",
    "            else:\n",
    "                s_fl, s_solo, s_tr = np.repeat(np.nan, 3)\n",
    "                \n",
    "            if og_contains_parent:\n",
    "                # get data on parental annotation\n",
    "                p_fl, p_solo, p_tr = rea_reindex.loc[(og, subg), ['fl', 'solo', 'tr']].iloc[0]\n",
    "                p_cn = og_cn.loc[(og, subg), 'cn']\n",
    "                \n",
    "                if has_annot:\n",
    "                    if has_assembly:\n",
    "                        if all([p_fl, s_solo]):\n",
    "                            cn = 0\n",
    "                            og_cn.loc[(og, s_subg), 'class'] = 'excision'\n",
    "                            \n",
    "                        elif all([p_solo, s_fl]):\n",
    "                            if og_fl_counts == '1=1':\n",
    "                                og_cn.loc[(og, s_subg), 'class'] = 'de_novo'\n",
    "                            else:\n",
    "                                og_cn.loc[(og, s_subg), 'class'] = 'inconsistent'\n",
    "                            \n",
    "                        elif all([p_fl, s_fl]):\n",
    "                            if cn == p_cn:\n",
    "                                og_cn.loc[(og, s_subg), 'class'] = 'no_change'\n",
    "                            else:\n",
    "                                og_cn.loc[(og, s_subg), 'class'] = 'aneup/LOH'\n",
    "                        \n",
    "                        elif all([p_solo, s_solo]) or all([p_tr, s_tr]):\n",
    "                            og_cn.loc[(og, s_subg), 'class'] = 'no_change'\n",
    "                                \n",
    "                        elif all([p_fl, s_tr]):\n",
    "                            og_cn.loc[(og, s_subg), 'class'] = 'truncation'\n",
    "                            \n",
    "                        else:\n",
    "                            og_cn.loc[(og, s_subg), 'class'] = 'Should_Not_Happen'\n",
    "                            \n",
    "                    else:\n",
    "                        og_cn.loc[(og, s_subg), 'class'] = 'inconsistent'\n",
    "                else:\n",
    "                    cn = 0\n",
    "                    if has_assembly:\n",
    "                        og_cn.loc[(og, s_subg), 'class'] = 'deletion'\n",
    "                        \n",
    "                    else:\n",
    "                        if p_fl:\n",
    "                            og_cn.loc[(og, s_subg), 'class'] = 'aneup/LOH'\n",
    "                        else:\n",
    "                            og_cn.loc[(og, s_subg), 'class'] = 'no_change'\n",
    "                        \n",
    "            else:\n",
    "                if has_annot:\n",
    "                    if has_assembly:\n",
    "                        # de novo\n",
    "                        if og_fl_counts == '1=1':\n",
    "                                og_cn.loc[(og, s_subg), 'class'] = 'de_novo'\n",
    "                        else:\n",
    "                            og_cn.loc[(og, s_subg), 'class'] = 'inconsistent'\n",
    "                            if og == 'og912':\n",
    "                                print('here')\n",
    "                        \n",
    "                    else:\n",
    "                        og_cn.loc[(og, s_subg), 'class'] = 'inconsistent'\n",
    "                        \n",
    "                else:\n",
    "                    og_cn.loc[(og, s_subg), 'class'] = 'empty'\n",
    "                    \n",
    "                    \n",
    "        # difference in copy number\n",
    "        delta_cn = cn-p_cn\n",
    "        og_cn.loc[(og, s_subg), 'delta_cn'] = delta_cn"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d289b17-6c26-488e-9b18-8cf3e59f90ff",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Deal with ogs calssified as complex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42615887-8f1e-457e-a7a7-29464b8885ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# export complex og\n",
    "#og_cn.loc[og_cn['class']=='complex'].to_csv('/mnt/HDD3/lrma/results/complex_og.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca74b3b8-a8c9-482c-8b5a-1952dc47384e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for og, df in og_cn.loc[(og_cn['class']=='complex')].reset_index().groupby('og'):\n",
    "    S = list(df['s_subg'].unique())\n",
    "    subg = S[0].split('.')[1]\n",
    "    if subg not in S:\n",
    "        S.append(subg)\n",
    "\n",
    "    dat = rea_reindex.loc[og].loc[S, ['family', 'Strand', 'fl', 'tr', 'solo','Start','End']].reset_index().sort_values(by=['s_subg', 'Start'])\n",
    "    fig, ax = plt.subplots(figsize=[6,3])\n",
    "    s_color_dict = dict(zip(S, sns.color_palette()))\n",
    "    annot = dat.apply(lambda x: [i for i in ['fl','solo','tr'] if x.loc[i]==True], axis=1)\n",
    "\n",
    "    for i in range(dat.shape[0]):\n",
    "        start, end, s_subg = dat.iloc[i][['Start', 'End', 's_subg']]\n",
    "        ax.plot([start, end], [i,i], c=s_color_dict[s_subg], label=s_subg)\n",
    "\n",
    "    ax.set_yticks(range(dat.shape[0]))\n",
    "    ax.set_yticklabels(annot)\n",
    "    plt.legend(bbox_to_anchor=[1,1], loc=2)\n",
    "    plt.tight_layout()\n",
    "\n",
    "    #plt.savefig(f'/mnt/HDD1/Dropbox/fig_lrma/{og}.complex.png', dpi=300)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e4427e3-edda-45c5-84f6-40635afb71a5",
   "metadata": {},
   "source": [
    "### Correct complex calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d37c6768-804a-4867-86d1-40c7613c6bda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# re-import and make corrections\n",
    "og_complex_corr = pd.read_csv('/mnt/HDD3/lrma/results/complex_og_corr.csv').set_index(['og','s_subg'])\n",
    "\n",
    "for i in og_complex_corr.index:\n",
    "    #add a safety switch to not override loci newly classified as non-complex\n",
    "    if og_cn.loc[i, 'class'] == 'complex':\n",
    "        og_cn.loc[i, 'class'] = og_complex_corr.loc[i, 'class']\n",
    "        og_cn.loc[i, 'delta_cn'] = og_complex_corr.loc[i, 'delta_cn']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8adb151-e0b4-42d7-ae06-d7db545e9b80",
   "metadata": {},
   "source": [
    "## Plot orthogroups contents and classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a24620b7-0fb1-4f9b-bbb6-8aa3cf765833",
   "metadata": {},
   "outputs": [],
   "source": [
    "for og in og_data.loc[og_data['og_fl']].index:\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=[7,7], gridspec_kw={'left':0.2, 'right':0.88, 'top':0.95, 'bottom':0.07})\n",
    "\n",
    "    S = og_cn.loc[og].index\n",
    "    X = og_data.loc[og, ['og_left_bound', 'og_right_bound']] + np.array([-1000, 1000])\n",
    "\n",
    "    for y, s in enumerate(S):\n",
    "\n",
    "        if all(og_cn.loc[(og, s), ['cons_as', 'has_assembly_left']]):\n",
    "            ax.axhline(y, color='k', lw=0.5, zorder=0)\n",
    "\n",
    "        if og_cn.loc[(og, s), 'cons_cn'] == True:\n",
    "            cn = og_cn.loc[(og, s), 'cn']\n",
    "            ax.text(X[0]+500, y, cn, color='k', va='center', size=8)\n",
    "\n",
    "        if og_cn.loc[(og, s), 'has_annot']:\n",
    "            rea = rea_reindex.loc[(og, s)].reset_index(drop=True)\n",
    "            for i in rea.index:\n",
    "                Start, End, Strand = rea.loc[i, ['Start', 'End', 'Strand']]\n",
    "                at = [j for j in ['fl', 'solo', 'tr'] if rea.loc[i, j]==True]\n",
    "                if len(at) == 1:\n",
    "                    at = at[0]\n",
    "                    fc = at_color[at]\n",
    "                else:\n",
    "                    fc = 'k'\n",
    "                fa = FancyArrow(Start, y, End-Start, 0, width=0.3, head_width=0.5, head_length=50, fc=fc, lw=0)\n",
    "                ax.add_patch(fa)\n",
    "\n",
    "        ax.text(X[1], y, og_cn.loc[(og, s), 'class'], va='center', size=8)\n",
    "\n",
    "    ax.set_yticks(range(len(S)))\n",
    "    ax.set_yticklabels(S, size=8)\n",
    "    ax.set_ylim(-1, len(S))\n",
    "    ax.set_xlim(X)\n",
    "    ax.set_title(og)\n",
    "\n",
    "    sns.despine()\n",
    "\n",
    "    #plt.savefig(f'{fig_path}/og_classification/{og}.png', dpi=300)\n",
    "    #plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45e20048-adb6-4c8a-9a0c-8de61d69a4e0",
   "metadata": {},
   "source": [
    "## Plot subgenome coverage with aneup/LOH calls to sort polyploidy, aneuploidy, and LOH events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c7a9768-9105-4960-8143-932cb3a86872",
   "metadata": {},
   "outputs": [],
   "source": [
    "Depth = {}\n",
    "\n",
    "for cross, df in ns_subg.loc[ns_subg['identity_filter']].groupby('cross'):\n",
    "\n",
    "    S_subg = Cross_parents[cross].split(',') + list(df.sort_values(['strain', 's_subg'])['s_subg'])\n",
    "    depth = []\n",
    "    for s_subg in S_subg:\n",
    "        d = pd.read_csv(f'/mnt/HDD3/lrma/depth_bins/{s_subg}.bins.tsv', sep='\\t', index_col=0)\n",
    "\n",
    "        if '.' not in s_subg:\n",
    "            s = 'parent'\n",
    "            subg = s_subg\n",
    "        else:\n",
    "            s, subg = s_subg.split('.')\n",
    "\n",
    "        d['Pos'] = d['Bin'] + tig_off[subg].loc[d['chrom'], 2].values\n",
    "        d['s_subg'] = s_subg\n",
    "        d['strain'] = s\n",
    "        d['subg'] = subg\n",
    "        depth.append(d)\n",
    "    depth = pd.concat(depth).reset_index(drop=True)\n",
    "    Depth[cross] = depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45b7c2f3-a573-4421-be6b-8b901a946c1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_cn_aneup_reindex = og_cn.loc[og_cn['class']=='aneup/LOH'].set_index('s_subg', drop=False)\n",
    "\n",
    "for c in ['cross', 'strain', 'subg']:\n",
    "    og_cn_aneup_reindex[c] = ns_subg.loc[og_cn_aneup_reindex['s_subg'], c].values\n",
    "for c in ['Query', 'og_mid']:\n",
    "    og_cn_aneup_reindex[c] = og_data.loc[og_cn_aneup_reindex['og'], c].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6e4be76-820f-4e55-9b2c-4d5300689f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "for cross in cross_order:\n",
    "    \n",
    "    depth = Depth[cross]\n",
    "    S_order = {j:i*8 for (i,j) in enumerate(['parent'] + list(nano_strains.loc[(nano_strains['identity_filter']) &\n",
    "                                                                             (nano_strains['cross']==cross), 'strain'].sort_values()))}\n",
    "\n",
    "    subg_order = dict(zip(Cross_parents[cross].split(','), [-1,1]))\n",
    "    subg_color = dict(zip(subg_order, ['blue', 'red']))\n",
    "    chrom_color = {0:'0.8', 1:'0.6'}\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=[10,10], gridspec_kw={'left':0.07, 'bottom':0.05, 'top':0.95, 'right':0.97})\n",
    "\n",
    "    for s, df in depth.sort_values(by='Pos').groupby('strain'):\n",
    "        y_order = S_order[s]\n",
    "\n",
    "        ax.axhline(y_order, c='k', lw=0.5, zorder=0)\n",
    "        for dy in [-3, -2, -1, 1, 2, 3]:\n",
    "            ax.axhline(y_order+dy, c='k', lw=0.5, ls=':', zorder=0)\n",
    "\n",
    "        for subg, df1 in df.groupby('subg'):\n",
    "            flip = subg_order[subg]\n",
    "            data_y = df1['depth']\n",
    "            data_y = (np.where(data_y>5, 5, data_y)*flip) + y_order\n",
    "            ax.plot(df1['Pos'], data_y, color=subg_color[subg], lw=0.5)\n",
    "\n",
    "            for chrom in tig_off[subg].index:\n",
    "                start, end, color = tig_off[subg].loc[chrom, [2,3,'color']]\n",
    "                fa = FancyArrow(start, y_order+flip, end-start, 0, width=2, head_length=0, color=chrom_color[color], lw=0, zorder=-1)\n",
    "                ax.add_artist(fa)\n",
    "\n",
    "            if s != 'parent':\n",
    "                s_subg = f'{s}.{subg}'\n",
    "                if s_subg in og_cn_aneup_reindex.index:\n",
    "                    og_sub = og_cn_aneup_reindex.loc[s_subg, 'og']\n",
    "                    og_sub_cn = og_cn_aneup_reindex.loc[s_subg, 'delta_cn']\n",
    "                    if type(og_sub) == str:\n",
    "                        og_sub = [og_sub]\n",
    "                        og_sub_cn = np.array([og_sub_cn])\n",
    "                    og_sub = og_data.loc[og_sub, ['og_mid', 'Query']]\n",
    "                    X = og_sub['og_mid']+tig_off[subg].loc[og_sub['Query'], 2].values\n",
    "                    Y = np.repeat(y_order+flip, og_sub.shape[0]) + og_sub_cn*flip\n",
    "                    ax.scatter(X, Y, c='#0f0f0f00', edgecolor=subg_color[subg], s=100, zorder=2)\n",
    "\n",
    "    y_ticks = sorted(S_order.values())\n",
    "    ax.set_yticks(y_ticks)\n",
    "    ax.set_yticklabels(pd.Series(S_order).sort_values().index)\n",
    "\n",
    "    legend_elms = [Line2D([0], [0], color=c, label=s) for (s,c) in subg_color.items()]\n",
    "    ax.legend(handles=legend_elms, ncol=2, loc='lower center', bbox_to_anchor=[0.5,1])\n",
    "\n",
    "    ax.margins(0.01)\n",
    "\n",
    "\n",
    "\n",
    "    #plt.savefig(f'{fig_path}aneupLOHog_depth_{cross}.png', dpi=300)\n",
    "    #plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b3c5fe9-50ed-469a-8e41-32430fad8e50",
   "metadata": {},
   "source": [
    "### Correct aneup/LOH calls\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "721f2c8b-fa4a-41a3-b2f4-85f9ac06db3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#og_cn_aneup_reindex.sort_values(['cross', 'strain', 'Query', 'og_mid'])\\\n",
    "#[['cross', 'strain', 'subg', 'Query', 'og_mid','og','delta_cn']].to_csv('/mnt/HDD3/lrma/results/aneupLOH_og.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1343755-6e97-4cdb-bb63-1ec07057262d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# og_corr was manually edited based on the coverage plots and parental orthologs\n",
    "# re-import and make corrections\n",
    "og_aneupLOH_corr = pd.read_csv('/mnt/HDD3/lrma/results/aneupLOH_og_corr.csv').set_index(['og','s_subg'])\n",
    "\n",
    "not_corr = []\n",
    "\n",
    "for i in og_aneupLOH_corr.index:\n",
    "    #add a safety switch to not override loci newly classified as non-aneup/LOH\n",
    "    if og_cn.loc[i, 'class'] == 'aneup/LOH':\n",
    "        og_cn.loc[i, 'class'] = og_aneupLOH_corr.loc[i, 'class']\n",
    "        og_cn.loc[i, 'delta_cn'] = og_aneupLOH_corr.loc[i, 'delta_cn']\n",
    "    else:\n",
    "        not_corr.append(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2d99dfd-20f1-4df3-b8cb-b86f1099c0c6",
   "metadata": {},
   "source": [
    "### Correct de novo calls based on manual inspection and TSD analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc8dbc66-6077-489c-b5e5-e0dec37e2391",
   "metadata": {},
   "outputs": [],
   "source": [
    "# From alignment inspection (Mauve)\n",
    "og_cn.loc[('og392','L41.LL2011_001'), 'class'] = 'inconsistent'\n",
    "og_cn.loc[('og5200','H33.UWOPS-91-202'), 'class'] = 'inconsistent'\n",
    "# From TSD sequence analysis (presence of TSD)\n",
    "og_cn.loc[('og5859','E30.YPS744'), 'class'] = 'inconsistent'\n",
    "og_cn.loc[('og1276','D68.LL2011_009'), 'class'] = 'inconsistent'\n",
    "# From TSD sequence analysis (non-uniqueness of TSD and surrounding sequence)\n",
    "og_cn.loc[('og1573', 'K47.LL2011_012'), 'class'] = 'inconsistent'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53a9ebef-eb15-4349-a165-ab2b6269ecf1",
   "metadata": {},
   "source": [
    "## Statistics on classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e40281f-b247-4de0-9ac8-60e2a4ad11f5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sub = og_cn.loc[og_cn['cross'].isin(cross_order)]\n",
    "100-sub.value_counts('class')/sub.shape[0]*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4ad78b5-e895-4c8e-8e85-80946012264a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sub = og_cn.loc[(og_cn['cross'].isin(cross_order)) & (og_cn['class']!='inconsistent')]\n",
    "sub.value_counts('class')/sub.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e23f6c5-18d9-4cd0-987d-e275956189a1",
   "metadata": {},
   "source": [
    "# Copy number changes per class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd82e36e-ac5b-440b-963d-733638a9fc03",
   "metadata": {},
   "outputs": [],
   "source": [
    "classes_order = ['polyploidy','LOH','aneuploidy','truncation','de_novo','deletion','excision']\n",
    "classes_alias = dict(zip(classes_order, ['Polyploidy','LOH','Aneuploidy','Truncation','De novo','Deletion','Excision']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38444be1-9322-44f3-b2a2-b52fa8ae2030",
   "metadata": {},
   "outputs": [],
   "source": [
    "Lines_per_cross = nano_strains.loc[nano_strains['identity_filter']].value_counts('cross')\n",
    "lines_per_subg = ns_subg.loc[ns_subg['identity_filter']].value_counts('subg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc2337e3-6f13-4288-bcb1-f9ce4a3094f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add variables to the og_cn table\n",
    "\n",
    "for (cross, subg), df in ns_subg.groupby(['cross', 'subg']):\n",
    "    S = df['s_subg'].values\n",
    "    sub = og_cn.loc[og_cn['s_subg'].isin(S)]\n",
    "    og_cn.loc[sub.index, 'cross'] = cross\n",
    "    og_cn.loc[sub.index, 'subg'] = subg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8280643-52a7-48f9-bc6b-24749060e7c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_CROSS = og_cn.value_counts(['cross','class'], sort=False).rename('count').reset_index()\n",
    "OG_CROSS['freq'] = OG_CROSS['count']/Lines_per_cross.loc[OG_CROSS['cross']].values\n",
    "\n",
    "OG_CROSS_CN = og_cn.value_counts(['cross','delta_cn','class'], sort=False).rename('count').reset_index()\n",
    "OG_CROSS_CN['freq'] = OG_CROSS_CN['count']/Lines_per_cross.loc[OG_CROSS_CN['cross']].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "941537b3-8b0d-4d55-a151-518024ba9e6e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "OG_STRAIN_CLASS = og_cn.loc[og_cn['class'].isin(classes_order)].reset_index(drop=True).copy()\n",
    "for c in ['strain', 'cross']:\n",
    "    OG_STRAIN_CLASS[c] = ns_subg.loc[OG_STRAIN_CLASS['s_subg'], c].values\n",
    "\n",
    "OG_STRAIN_CLASS = OG_STRAIN_CLASS.value_counts(['cross','class','strain']).rename('count')\n",
    "\n",
    "for cross in cross_order:\n",
    "    lc = lines_per_cross[cross]\n",
    "    \n",
    "    for c in classes_order:\n",
    "        \n",
    "        if (cross, c) in OG_STRAIN_CLASS.index:\n",
    "            sub = OG_STRAIN_CLASS.loc[(cross, c)]\n",
    "            for s in lc:\n",
    "                if s not in sub.index:\n",
    "                    OG_STRAIN_CLASS.loc[(cross, c, s)] = 0\n",
    "                    \n",
    "        else:\n",
    "            for s in lc:\n",
    "                OG_STRAIN_CLASS.loc[(cross, c, s)] = 0\n",
    "                \n",
    "OG_STRAIN_CLASS = OG_STRAIN_CLASS.reset_index()\n",
    "OG_STRAIN_CLASS['class_order'] = OG_STRAIN_CLASS['class'].replace(dict(zip(classes_order, range(7))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e12e322a-3fd5-4ae7-8bfc-cdb0aa2b284a",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_CCC = og_cn.loc[og_cn['class']!='parental'].reset_index(drop=True).copy()\n",
    "for c in ['strain', 'cross']:\n",
    "    OG_CCC[c] = ns_subg.loc[OG_CCC['s_subg'], c].values\n",
    "\n",
    "for delta_cn, df in OG_CCC.groupby('delta_cn'):\n",
    "    if delta_cn < 0:\n",
    "        OG_CCC.loc[df.index, 'delta_cn_direction'] = 'loss'\n",
    "    elif delta_cn > 0:\n",
    "        OG_CCC.loc[df.index, 'delta_cn_direction'] = 'gain'\n",
    "\n",
    "OG_CCC = OG_CCC.groupby(['class','cross','delta_cn_direction'])['delta_cn'].sum().rename('delta_cn').reset_index()\n",
    "OG_CCC['delta_cn_per_line'] = OG_CCC['delta_cn']/Lines_per_cross.loc[OG_CCC['cross']].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21ff1b26-10ff-45b2-8e44-243c2c3a75a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_STRAIN_CN = og_cn.loc[og_cn['class']!='parental'].reset_index(drop=True).copy()\n",
    "for c in ['strain', 'cross']:\n",
    "    OG_STRAIN_CN[c] = ns_subg.loc[OG_STRAIN_CN['s_subg'], c].values\n",
    "    \n",
    "OG_STRAIN_CN = OG_STRAIN_CN.groupby(['strain','cross'])['delta_cn'].sum().rename('delta_cn').reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50117bb4-3e70-4631-a32a-2ca8241e675a",
   "metadata": {},
   "outputs": [],
   "source": [
    "wsr_delta_cn = []\n",
    "\n",
    "for cross, df in OG_STRAIN_CN.groupby('cross'):\n",
    "    \n",
    "    w, p = stats.wilcoxon(df['delta_cn'])\n",
    "    wsr_delta_cn.append([cross, w, p])\n",
    "    \n",
    "wsr_delta_cn = pd.DataFrame(wsr_delta_cn, columns=['cross', 'W', 'pval']).set_index('cross')\n",
    "wsr_delta_cn['pval_corr'] = multipletests(wsr_delta_cn['pval'], method='fdr_bh')[1]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f81c6a8-6519-444c-b5e4-63141e49c68b",
   "metadata": {},
   "source": [
    "## Fig 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffa85c2a-ed02-4158-8e1c-722a008e58d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=[8, 6])\n",
    "\n",
    "gs = plt.GridSpec(ncols=2, nrows=2, height_ratios=[4,3], width_ratios=[3,4], wspace=0.3, hspace=0.3, \n",
    "                 left=0.09, right=0.98, top=0.88, bottom=0.09)\n",
    "\n",
    "gs1 = plt.GridSpec(ncols=7, nrows=2, height_ratios=[4,3], wspace=0.35, hspace=0.3, \n",
    "                 left=0.09, right=0.98, top=0.88, bottom=0.09)\n",
    "\n",
    "\n",
    "## Fig 3A: counts of events per line per class\n",
    "\n",
    "classes_dx = dict(zip(classes_order, range(7)))\n",
    "\n",
    "w = 0.7\n",
    "\n",
    "for c, df in OG_STRAIN_CLASS.loc[OG_STRAIN_CLASS['class'].isin(classes_order)].groupby('class'):\n",
    "        \n",
    "    ax = fig.add_subplot(gs1[0, classes_dx[c]])\n",
    "\n",
    "    df1 = df.groupby('cross')['count'].mean()\n",
    "    \n",
    "    \n",
    "    ax.bar(range(11), df1.loc[list(cross_order)], color=[cross_color[cross] for cross in cross_order])\n",
    "    \n",
    "    sns.stripplot(x='cross', y='count', order=cross_order, size=3, color='black', alpha=0.3,\n",
    "                  jitter=0.15, clip_on=False, data=df, ax=ax)\n",
    "    \n",
    "    ax.set_xlim(-1, 11)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_xlabel('Cross')\n",
    "    \n",
    "    if classes_dx[c] != 0:\n",
    "        ax.set_ylabel('')\n",
    "    else:\n",
    "        ax.set_ylabel('Ty loci per line')\n",
    "\n",
    "    if df['count'].max()>10:\n",
    "        ax.set_ylim(0, 40)\n",
    "        ax.set_yticks(range(0,41,10))\n",
    "        \n",
    "    elif df['count'].max()>4:\n",
    "        ax.set_ylim(0, 10)\n",
    "        ax.set_yticks(range(0,11,2))\n",
    "        \n",
    "    elif df['count'].max()>2:\n",
    "        ax.set_ylim(0, 4)\n",
    "        ax.set_yticks(range(0,5,1))\n",
    "    \n",
    "    else:\n",
    "        ax.set_ylim(0, 2)\n",
    "        ax.set_yticks(range(0,3,1))\n",
    "        \n",
    "    ax.text(0.5, 1.05, classes_alias[c], size=12, va='bottom', ha='center', transform=ax.transAxes)\n",
    "    \n",
    "ax_legend = fig.add_axes([0, 1, 1, 0])\n",
    "ax_legend.axis('off')\n",
    "\n",
    "legend_elms = [Line2D([0], [0], color='w', marker='s', ms=9, mfc=cross_color[cross], label=cross) for cross in cross_order]\n",
    "ax_legend.legend(handles=legend_elms, loc='upper right', bbox_to_anchor=[1,1], ncol=11,\n",
    "                 columnspacing=1.5, handlelength=0.5, labelspacing=0.3)\n",
    "\n",
    "## Fig 3B: plot the delta CN per cross and per mechanism\n",
    "\n",
    "ax = fig.add_subplot(gs[1, 0])\n",
    "\n",
    "classes_dx = {'LOH':0, 'aneuploidy':1}\n",
    "\n",
    "w = 0.06\n",
    "\n",
    "for (cross, c, d), df in OG_CCC.loc[OG_CCC['class'].isin(classes_dx)].groupby(['cross', 'class', 'delta_cn_direction']):\n",
    "    \n",
    "    x = classes_dx[c] + (cross_order[cross]-5)/15\n",
    "    dy = df['delta_cn_per_line'].iloc[0]\n",
    "    \n",
    "    fa = FancyArrow(x, 0, 0, dy, fc=cross_color[cross], lw=0, head_length=0, head_width=0, width=w, clip_on=False, zorder=1)\n",
    "    ax.add_patch(fa)\n",
    "\n",
    "ax.axhline(0, ls='--', c='k', zorder=-1)\n",
    "\n",
    "ax.set_xlim(-0.5, 1.5)\n",
    "ax.set_xticks([0,1])\n",
    "ax.set_xticklabels([classes_alias[c] for c in classes_dx])\n",
    "\n",
    "ax.set_ylim(-3, 3)\n",
    "ax.set_ylabel('CN gain/loss per line')\n",
    "\n",
    "\n",
    "## Fig 3C: plot the distribution of cn changes per strain\n",
    "\n",
    "ax = fig.add_subplot(gs[1, 1])\n",
    "\n",
    "sns.boxplot(x='cross', y='delta_cn', data=OG_STRAIN_CN, color='w', fliersize=0,\n",
    "            order=cross_order, ax=ax)\n",
    "\n",
    "sns.stripplot(x='cross', y='delta_cn', data=OG_STRAIN_CN, palette=cross_color, alpha=0.5, hue='cross',\n",
    "            order=cross_order, ax=ax)\n",
    "\n",
    "# plot p-values\n",
    "for cross, df in OG_STRAIN_CN.groupby('cross'):\n",
    "    y = df['delta_cn'].max()+2\n",
    "    x = cross_order[cross]\n",
    "    \n",
    "    ax.text(x, y, plot_pval_symbol(wsr_delta_cn.loc[cross, 'pval_corr']), ha='center', va='bottom')\n",
    "\n",
    "ax.axhline(0, ls='--', c='k', zorder=-1)\n",
    "ax.legend_.remove()\n",
    "ax.set_ylabel('Net CN change per line')\n",
    "ax.set_xticklabels(cross_order, rotation=90)\n",
    "ax.set_xlabel('')\n",
    "\n",
    "fig.text(0.01, 0.95, 'A', size=24, fontweight='bold', font='Arial')\n",
    "fig.text(0.01, 0.40, 'B', size=24, fontweight='bold', font='Arial')\n",
    "fig.text(0.45, 0.40, 'C', size=24, fontweight='bold', font='Arial')\n",
    "\n",
    "\n",
    "sns.despine()\n",
    "\n",
    "#for ext in ['svg', 'png']:\n",
    "#    plt.savefig(f'{fig_path}Fig3.{ext}', dpi=300)\n",
    "\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a005bedd-5e26-4796-a1d0-4f2a0d2cbe2e",
   "metadata": {},
   "source": [
    "## Classification failure\n",
    "## Fig S3B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f51b0be7-82aa-4e29-bce6-83072477ca52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import assembl stats\n",
    "assembl_stats = pd.read_csv(f'{tables_path}assembly_stats.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11ce22af-8347-491a-8410-f18968fbb6aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# percentage of inconsistently classified orthogroups\n",
    "perc_inconsistent = og_cn.reset_index(drop=True).value_counts(['s_subg','class']).rename('count').reset_index()\n",
    "perc_inconsistent = pd.pivot_table(index='s_subg', columns='class', values='count', data=perc_inconsistent).fillna(0)\n",
    "perc_inconsistent['sum'] = perc_inconsistent.sum(axis=1)\n",
    "perc_inconsistent = perc_inconsistent.apply(lambda x: x/perc_inconsistent['sum']*100, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97230008-5bd0-4a52-9952-ac4cb775f0db",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=[4.5,4])\n",
    "dat = pd.concat([perc_inconsistent['inconsistent'], assembl_stats[['cross','n']]], axis=1)\n",
    "sns.scatterplot(x='n', y='inconsistent', hue='cross', palette=cross_color,\n",
    "                hue_order=cross_order, data=dat, ax=ax)\n",
    "\n",
    "ax.set_xlabel('Number of contigs')\n",
    "ax.set_ylabel('Classification failure (%)')\n",
    "ax.legend(loc=6, bbox_to_anchor=[1,0,0,1])\n",
    "\n",
    "fig.text(0, 0.92, 'B', size=24, fontweight='bold', font='Arial')\n",
    "\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "\n",
    "#plt.show()\n",
    "#plt.savefig(f'{fig_path}Supp_FigS3B.svg')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbfcb734-83c3-48bf-9320-61121d6c66a9",
   "metadata": {},
   "source": [
    "# Correlation with the previous short-read estimates of copy numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0b92032-18eb-4527-9780-c2352e7603c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_STRAIN_FAM_CN = og_cn.loc[og_cn['class'].isin(classes_order+['no_change'])].reset_index(drop=True).copy()\n",
    "for c in ['strain', 'cross']:\n",
    "    OG_STRAIN_FAM_CN[c] = ns_subg.loc[OG_STRAIN_FAM_CN['s_subg'], c].values\n",
    "\n",
    "OG_STRAIN_FAM_CN = OG_STRAIN_FAM_CN.value_counts(['og','cross','strain','delta_cn']).rename('count').reset_index()\n",
    "# add orthogroup family\n",
    "OG_STRAIN_FAM_CN['family'] = og_data.loc[OG_STRAIN_FAM_CN['og'], 'family'].values\n",
    "OG_STRAIN_FAM_CN = OG_STRAIN_FAM_CN.groupby(['cross','strain','family'])['delta_cn'].sum().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5147d3a-7ba6-42de-b9e4-45d2484e8e02",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_ELIFE = OG_STRAIN_FAM_CN.loc[OG_STRAIN_FAM_CN['family'].isin(['Ty1p', 'TY1', 'Ty3p', 'Tsu4'])].copy()\n",
    "OG_ELIFE['Fam'] = OG_ELIFE['family'].replace({'Ty1p':'Ty1', 'TY1':'Ty1'})\n",
    "\n",
    "# collapse Ty1 counts\n",
    "OG_ELIFE = OG_ELIFE.groupby(['strain','cross','Fam'])['delta_cn'].sum().reset_index()\n",
    "\n",
    "# add elife counts\n",
    "ty_cn_elife = pd.read_excel('/mnt/HDD3/lrma/misc/elife-60474-supp1-v2.xlsx', sheet_name='Supplementary_File_1f')\n",
    "ty_cn_elife = ty_cn_elife.loc[ty_cn_elife['Timepoint']=='Tend'].set_index('Strain', drop=False)\n",
    "\n",
    "for (s, fam), df in OG_ELIFE.groupby(['strain','Fam']):\n",
    "    if s in ty_cn_elife.index:\n",
    "        cn_elife = ty_cn_elife.loc[s, fam]\n",
    "        OG_ELIFE.loc[df.index, 'cn_elife'] = 2**cn_elife\n",
    "OG_ELIFE.dropna(axis=0, subset='cn_elife', inplace=True)\n",
    "\n",
    "OG_ELIFE['cross_Fam'] = OG_ELIFE.apply(lambda x: f'{x[\"cross\"]}.{x[\"Fam\"]}', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7549e3cb-7181-4f6d-b4fe-c0bf9fa6ef30",
   "metadata": {},
   "outputs": [],
   "source": [
    "m = smf.mixedlm(\"cn_elife ~ delta_cn\", OG_ELIFE, groups=OG_ELIFE['cross_Fam'])\n",
    "mf = m.fit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51572992-ffcb-4f5b-92c9-3310b8270881",
   "metadata": {},
   "outputs": [],
   "source": [
    "mf.pvalues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80526659-860c-40db-a05d-b942a906c52a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mf.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bac9ae1-77d0-4f39-bbb4-e3e4f38d8ef4",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_ELIFE_REG = []\n",
    "for (cross, fam), df in OG_ELIFE.groupby(['cross','Fam']):\n",
    "    X = df['delta_cn']\n",
    "    Y = df['cn_elife']\n",
    "    if X.unique().shape[0] > 1:\n",
    "        lm = stats.linregress(X, Y)\n",
    "        spearman = stats.spearmanr(X, Y)\n",
    "        OG_ELIFE_REG.append([cross, fam, lm.slope, lm.intercept, lm.rvalue, lm.pvalue, spearman.pvalue])\n",
    "        \n",
    "OG_ELIFE_REG = pd.DataFrame(OG_ELIFE_REG, columns=['cross', 'Fam', 'slope', 'intercept', 'rval', 'pval', 'pval_spearman'])\n",
    "OG_ELIFE_REG['pval_corr'] = multipletests(OG_ELIFE_REG['pval'], method='fdr_bh')[1]\n",
    "OG_ELIFE_REG.set_index(['cross', 'Fam'], inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6a7c7bf-cafb-45ff-8413-47545f4dd327",
   "metadata": {},
   "source": [
    "## Fig S6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5244733-878b-4831-99a1-262a105b4815",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=[10, 9])\n",
    "gs = plt.GridSpec(ncols=6, nrows=6, hspace=0.8, wspace=0.7,\n",
    "                 left=0.08, right=0.97, top=0.95, bottom=0.07)\n",
    "\n",
    "fam_axes = dict(zip(['Ty1', 'Ty3p', 'Tsu4'], range(3)))\n",
    "\n",
    "for (cross, fam), df in OG_ELIFE.groupby(['cross', 'Fam']):\n",
    "\n",
    "    ax_idx = np.array([fam_axes[fam], cross_order[cross]])\n",
    "    if cross in ['BC1', 'BC2', 'BA1', 'BA2', 'BSc1', 'BSc2']:\n",
    "        ax_idx = ax_idx + np.array([3,-5])\n",
    "    \n",
    "    ax = fig.add_subplot(gs[tuple(ax_idx)])\n",
    "    ax.scatter(df['delta_cn'], df['cn_elife'], color=cross_color[cross], zorder=1, clip_on=False)\n",
    "    \n",
    "    # plot regression\n",
    "    \n",
    "    if (cross, fam) in OG_ELIFE_REG.index:\n",
    "        pval, rval, slope, intercept = OG_ELIFE_REG.loc[(cross, fam), ['pval_corr', 'rval', 'slope', 'intercept']]\n",
    "        X = np.array([df['delta_cn'].min(), df['delta_cn'].max()])\n",
    "        ax.plot(X, (X*slope)+intercept, c='k', zorder=0)\n",
    "        if pval < 0.05:\n",
    "            fw = 'bold'\n",
    "        else:\n",
    "            fw = 'normal'\n",
    "        ax.set_title(f'{cross} {fam}\\np={plot_pval_text(pval)}, r={rval:.2f}', size=10, fontweight=fw)\n",
    "    else:\n",
    "        ax.set_title(f'{cross} {fam}', size=10)\n",
    "        \n",
    "    \n",
    "fig.text(0.02, 0.5, 'CN (short-read relative depth of coverage)', rotation=90, size=18, ha='center', va='center')\n",
    "fig.text(0.5, 0.02, 'CN change (subgenome assemblies)', size=18, ha='center', va='center')\n",
    "\n",
    "sns.despine()\n",
    "\n",
    "#for ext in ['svg', 'png']:\n",
    "#    plt.savefig(f'{fig_path}Supp_FigS6.{ext}', dpi=300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "464b2c3b-30fe-46d3-94e5-5227505ecf58",
   "metadata": {},
   "source": [
    "# De novo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca74e238-83e8-4525-9b16-bcf9d2491c96",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_denovo = og_cn.loc[og_cn['class']=='de_novo'].copy().reset_index(drop=True)\n",
    "for s_subg, df in og_denovo.groupby('s_subg'):\n",
    "    if s_subg in ns_subg.index:\n",
    "        og_denovo.loc[df.index, 'cross'] = ns_subg.loc[s_subg, 'cross']\n",
    "og_denovo['cross_order'] = og_denovo['cross'].apply(lambda x: cross_order[x])\n",
    "og_denovo = og_denovo.sort_values(by=['cross_order', 's_subg'])\n",
    "\n",
    "for (s_subg, og), df in og_denovo.groupby(['s_subg', 'og']):\n",
    "    \n",
    "    rea_sub = rea_reindex.loc[(og, s_subg)]\n",
    "    rea_sub = rea_sub.loc[rea_sub['fl']==True].iloc[0]\n",
    "    \n",
    "    for c in ['id', 'family', 'Query', 'Start', 'End', 'Strand', 'query', 'start', 'end']:\n",
    "        og_denovo.loc[df.index, c] = rea_sub.loc[c]\n",
    "    \n",
    "    og_denovo.loc[df.index, 'og_complex'] = og_data.loc[og, 'og_complex']\n",
    "    \n",
    "og_denovo['denovo_size'] = og_denovo['end']-og_denovo['start']\n",
    "og_denovo['denovo_size_ref'] = og_denovo['End']-og_denovo['Start']\n",
    "# manually add strand data for one de novo og\n",
    "og_denovo.loc[(og_denovo['s_subg']=='D12.UWOPS-91-202') & (og_denovo['og']=='og5121'), 'Strand'] = '-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28c648a9-6e27-4901-b916-b67cf48d78f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# export og_denovo\n",
    "#og_denovo.to_csv('/mnt/HDD3/lrma/results/og_denovo.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f294bbe8-9070-4f2c-870e-f20593d62cac",
   "metadata": {},
   "source": [
    "## Disruptive de novo insertions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65f29122-fe8a-47e7-b422-3e887e581293",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (subg, chrom), df in og_denovo.groupby(['subg', 'Query']):\n",
    "    assembly = assembly_background_conversion[subg]\n",
    "    \n",
    "    gff = GFF[assembly]\n",
    "    gff = gff.loc[gff[0]==chrom]\n",
    "    t = intervaltree.IntervalTree()\n",
    "    for j in gff.index:\n",
    "        start, end = gff.loc[j, [3,4]].values\n",
    "        t[start:end] = j\n",
    "\n",
    "    for i in df.index:\n",
    "        start, end = og_denovo.loc[i, ['Start', 'End']].sort_values()\n",
    "        x = t[start:end]\n",
    "        \n",
    "        if len(x) > 0:\n",
    "            print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5f74ae4-5bcb-48af-a615-c9f40c7e4533",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import tblastn results for AIM29\n",
    "aim29_tblastn = []\n",
    "\n",
    "for s_subg in ns_subg.loc[(ns_subg['subg']=='LL2011_009') & (ns_subg['identity_filter']), 's_subg']:\n",
    "    tab = pd.read_csv(f'/mnt/HDD3/lrma/sequences/AIM29.{s_subg}.tblastn.tab', sep='\\t', header=None)\n",
    "    tab = tab.loc[tab[1]==tab[1].iloc[0]]\n",
    "    tab['s_subg'] = s_subg\n",
    "    \n",
    "    aim29_tblastn.append(tab)\n",
    "    \n",
    "aim29_tblastn = pd.concat(aim29_tblastn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e0e6155-739e-41e0-9368-c167cf6c7d6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export corresponding genomic sequences for alignment\n",
    "aim29_seq = []\n",
    "\n",
    "for s_subg, df in aim29_tblastn.groupby('s_subg'):\n",
    "    \n",
    "    genome = {seq.id:seq for seq in SeqIO.parse(f'/mnt/HDD3/lrma/medaka/{s_subg}/consensus.fasta', 'fasta')}\n",
    "    \n",
    "    chrom = df[1].iloc[0]\n",
    "    pos = df[[8,9]].values.flatten()\n",
    "    \n",
    "    if s_subg == 'L41.LL2011_009':\n",
    "        seq = genome[chrom].seq[pos.min()-20-1:pos.max()+20+6000]\n",
    "    \n",
    "    else:\n",
    "        seq = genome[chrom].seq[pos.min()-20-1:pos.max()+20]\n",
    "    \n",
    "    if df.iloc[0, 9] < df.iloc[0, 8]:\n",
    "        seq = seq.reverse_complement()\n",
    "        \n",
    "    sr = SeqRecord.SeqRecord(seq=seq, id=f'AIM29-{s_subg}', description='')\n",
    "    \n",
    "    aim29_seq.append(sr)\n",
    "    \n",
    "with open('/mnt/HDD3/lrma/sequences/all.AIM29.fasta', 'w') as handle:\n",
    "    SeqIO.write(aim29_seq, handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ad84172-ffd1-4afd-80c2-5ed81b607526",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_fl_annot(rea_sub, z=0, w=1, y=0):\n",
    "    \n",
    "    global ax\n",
    "    \n",
    "    ty_elm_palette = {'Ty1p': {'LTR':'midnightblue', 'I':'dodgerblue'}, 'Tsu4':{'LTR':'darkred', 'I':'lightcoral'}}\n",
    "    ty_elm_hl = {'LTR':100, 'I':0}\n",
    "    ty_elm_zdelta = {'LTR':2, 'I':1}\n",
    "    \n",
    "    start, end1, start2, end, strand, fam = rea_sub.loc[['start','end1','start2','end','orient','family']]\n",
    "\n",
    "    for elm, (c1,c2) in zip(['LTR', 'I', 'LTR'], [(start, end1), (start, end), (start2, end)]):\n",
    "        if strand == 'C':\n",
    "            (c1, c2) = (c2, c1)\n",
    "\n",
    "        zdelta = ty_elm_zdelta[elm]\n",
    "        hl = ty_elm_hl[elm]\n",
    "        c = ty_elm_palette[fam][elm]\n",
    "\n",
    "        fa = FancyArrow(c1, y, (c2-c1), 0, color=c, lw=0, zorder=z+zdelta, clip_on=False,\n",
    "                        head_length=hl, width=w, head_width=w, length_includes_head=True)\n",
    "        ax.add_patch(fa)\n",
    "        \n",
    "    fa = FancyArrow(start, y, (end-start), 0, color='w', lw=0, zorder=z, clip_on=False,\n",
    "                        head_length=0, width=2, head_width=2, length_includes_head=True)\n",
    "    ax.add_patch(fa)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf985605-0d28-4154-8a78-6ce162efeabb",
   "metadata": {},
   "source": [
    "## Fig S10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "458cdb76-4590-41ba-b2c7-e1ae7e562566",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(nrows=2, figsize=[7,3])\n",
    "\n",
    "# Nested insertion in J27.MSH-587-1\n",
    "\n",
    "ax = axes[0]\n",
    "\n",
    "rea_sub = rea_reindex.loc[('og4121', 'J27.MSH-587-1')].set_index('id')\n",
    "\n",
    "plot_fl_annot(rea_sub.loc['i3'], z=0, w=1.5)\n",
    "plot_fl_annot(rea_sub.loc['i4'], z=2, w=1)\n",
    "\n",
    "ax.axhline(0, lw=2, c='k', zorder=-1)\n",
    "\n",
    "ax.set_xlim(np.array([-7500, 7500])+126500)\n",
    "ax.set_ylim(-1.2, 1.2) \n",
    "ax.set_yticks([])\n",
    "ax.set_xlabel('chromosome XVI')\n",
    "ax.set_title('J27.MSH-587-1 (CC1)')\n",
    "\n",
    "for s in ['top', 'left', 'right']:\n",
    "    ax.spines[s].set_visible(False)\n",
    "    \n",
    "# insertion in AIM29\n",
    "\n",
    "ax = axes[1]\n",
    "rea_sub = rea_reindex.loc['og1236'].iloc[0]\n",
    "plot_fl_annot(rea_sub, z=2)\n",
    "\n",
    "# coordinates of AIM29 on ctg6 of L41.LL2011_009\n",
    "end_aim29 = 76274\n",
    "start_aim29 = 82594+40\n",
    "w = 1.5\n",
    "fa = FancyArrow(start_aim29, 0, end_aim29-start_aim29, 0, color='k', lw=0, zorder=0, clip_on=False,\n",
    "                width=w, head_width=w, head_length=100, length_includes_head=True)\n",
    "ax.add_patch(fa)\n",
    "\n",
    "ax.axhline(0, lw=2, c='k', zorder=-1)\n",
    "    \n",
    "ax.set_xlim(np.array([7500, -7500])+79500)\n",
    "ax.set_ylim(-1.2, 1.2) \n",
    "ax.set_yticks([])\n",
    "ax.set_xlabel('chromosome XI')\n",
    "ax.set_title('$AIM29$ L41.LL2011_009 (CC3)')\n",
    "\n",
    "for s in ['top', 'left', 'right']:\n",
    "    ax.spines[s].set_visible(False)\n",
    "\n",
    "\n",
    "\n",
    "fig.text(0.01, 0.90, 'A', size=24, fontweight='bold', font='Arial')\n",
    "fig.text(0.01, 0.40, 'B', size=24, fontweight='bold', font='Arial')\n",
    "    \n",
    "plt.tight_layout()\n",
    "\n",
    "#for ext in ['png', 'svg']:\n",
    "#    plt.savefig(f'{fig_path}Supp_FigS10.{ext}', dpi=300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10147092-fb39-48de-a569-9c0121aee815",
   "metadata": {},
   "source": [
    "## Plot de novo insertion loci\n",
    "\n",
    "## Fig 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c6fe0d6-64f0-447d-9eb5-82fd4e28d3dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare a dataframe of lines with de novo insertions\n",
    "lines_denovo = ns_subg.loc[ns_subg['s_subg'].isin(og_denovo['s_subg'].unique()), 'strain'].unique()\n",
    "lines_denovo = ns_subg.loc[ns_subg['strain'].isin(lines_denovo)].copy()\n",
    "lines_denovo['strain_order'] = lines_denovo['strain'].apply(lambda x: int(x[1:]))\n",
    "lines_denovo['cross_order'] = lines_denovo['cross'].apply(lambda x: cross_order[x])\n",
    "\n",
    "for cross, df in lines_denovo.groupby('cross'):\n",
    "    p1, p2 = Cross_parents[cross].split(',')\n",
    "    subg_idx = pd.Series({p1:0, p2:1})\n",
    "    lines_denovo.loc[df.index, 'subg_order'] = subg_idx.loc[df['subg']].values\n",
    "    \n",
    "# final sorting\n",
    "lines_denovo = lines_denovo.sort_values(by=['cross_order', 'strain_order','subg_order']).set_index('s_subg')\n",
    "lines_denovo['order'] = range(lines_denovo.shape[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5237aa5a-5e35-45ba-a824-2e2b4fcee1b1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "og_cn_lines_denovo = og_cn.reset_index(drop=True)\n",
    "og_cn_lines_denovo = og_cn_lines_denovo.loc[(og_cn_lines_denovo['s_subg'].isin(lines_denovo.index)) &\n",
    "                                           (og_cn_lines_denovo['has_fl'])].copy()\n",
    "og_cn_lines_denovo['CN'] = og_cn_lines_denovo['cn']+og_cn_lines_denovo['delta_cn']\n",
    "og_cn_lines_denovo = og_cn_lines_denovo.loc[(og_cn_lines_denovo['class'].isin(['LOH','aneuploidy','polyploidy','de_novo','no_change'])) & \n",
    "                                           (og_cn_lines_denovo['CN']>=1)]\n",
    "\n",
    "# add OG variables as in og_denovo\n",
    "for (s_subg, og), df in og_cn_lines_denovo.groupby(['s_subg', 'og']):\n",
    "    \n",
    "    rea_sub = rea_reindex.loc[(og, s_subg)]\n",
    "    rea_sub = rea_sub.loc[rea_sub['fl']==True].iloc[0]\n",
    "    \n",
    "    for c in ['id', 'family', 'Query', 'Start', 'End', 'Strand', 'query', 'start', 'end']:\n",
    "        og_cn_lines_denovo.loc[df.index, c] = rea_sub.loc[c]\n",
    "    \n",
    "    og_cn_lines_denovo.loc[df.index, 'og_complex'] = og_data.loc[og, 'og_complex']\n",
    "    \n",
    "# manually add strand data for one de novo og\n",
    "og_cn_lines_denovo.loc[(og_cn_lines_denovo['s_subg']=='D12.UWOPS-91-202') & (og_cn_lines_denovo['og']=='og5121'), 'Strand'] = '-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f85c5331-f3b0-4f63-bc9a-8d57443a64e4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=[8, 7], \n",
    "                       gridspec_kw={'top':0.93, 'bottom':0.07, 'right':0.9, 'left':0.17})\n",
    "\n",
    "chrom_color = {0:'0.75', 1:'0.90'}\n",
    "strand_dy = {'+':+0.15, '-':-0.15}\n",
    "strand_marker = {'+': '>', '-': '<'}\n",
    "cross_order_denovo = ['CC1', 'CC2', 'CC3', 'BB1', 'BC2', 'BA1', 'BA2', 'BSc1']\n",
    "agc = 4\n",
    "gap_crosses = dict(zip(cross_order_denovo, np.linspace(0, agc, 8)))\n",
    "y_tick_labels = {}\n",
    "\n",
    "denovo_counts_per_cross = og_denovo.value_counts('cross')\n",
    "   \n",
    "for s_subg in lines_denovo.index:\n",
    "    \n",
    "    df = og_cn_lines_denovo.loc[og_cn_lines_denovo['s_subg']==s_subg]\n",
    "    \n",
    "    Y, cross = lines_denovo.loc[s_subg, ['order', 'cross']]\n",
    "    if Y%2 == 1:\n",
    "        y = Y-0.1\n",
    "    else:\n",
    "        y = Y+0.1\n",
    "        \n",
    "    # Add cross-specific spacing\n",
    "    y += gap_crosses[cross]\n",
    "    \n",
    "    y_tick_labels[s_subg] = y\n",
    "    \n",
    "    s, subg = s_subg.split('.')\n",
    "    \n",
    "    if df.shape[0]>0:\n",
    "        \n",
    "        for i in df.index:\n",
    "            fam, Chrom, Start, End, Strand, c = og_cn_lines_denovo.loc[i, ['family', 'Query', 'Start', 'End', 'Strand', 'class']]\n",
    "\n",
    "            if c == 'de_novo':\n",
    "                mfc = ty_colors[fam]\n",
    "                mec = ty_colors[fam]\n",
    "                size = 36\n",
    "\n",
    "            else:\n",
    "                mfc = (1,1,1,0)\n",
    "                mec = ty_colors[fam]\n",
    "                size = 24\n",
    "\n",
    "            ax.scatter(np.mean([Start,End])+tig_off[subg].loc[Chrom, 2], y+strand_dy[Strand], zorder=2, \n",
    "                        marker=strand_marker[Strand], color=mfc, edgecolors=mec, lw=0.7, s=size)\n",
    "\n",
    "    # add chromosome paintings\n",
    "    for (chrom, length, start, color), df in tig_off[subg].groupby([0,1,2,'color']):\n",
    "\n",
    "        ar = FancyArrow(start, y, length, 0, width=0.5, head_width=1, zorder=0,\n",
    "                        fc=chrom_color[color], lw=0, length_includes_head=True, clip_on=False, head_length=0)\n",
    "        ax.add_patch(ar)\n",
    "\n",
    "        if s_subg == 'C49.LL2013_040':\n",
    "            ax.text(start+0.5*length, y+0.8, chrom[3:], ha='center', va='center', size=8, color='k', zorder=1)\n",
    "\n",
    "\n",
    "y_tick_labels = pd.Series(y_tick_labels).sort_values()\n",
    "ax.set_yticks(y_tick_labels.values)\n",
    "ax.set_yticklabels(y_tick_labels.index, size=8)\n",
    "ax.set_ylim(-1, 32+agc)\n",
    "\n",
    "ax.set_xlim(-1e5, 12e6)\n",
    "ax.set_xticks(np.arange(0, 12.1e6, 1e6))\n",
    "ax.set_xticklabels(range(13))\n",
    "ax.set_xlabel('Mb')\n",
    "\n",
    "# add patches for crosses\n",
    "for cross, dy in zip(cross_order_denovo, \n",
    "                     [(0,7), (8,11), (12,15), (16,17), (18,23), (24,25), (26,29), (30,31)]):\n",
    "    \n",
    "    y = np.mean(dy) + gap_crosses[cross]\n",
    "    if cross == 'BSc1':\n",
    "        tc = 'k'\n",
    "    else:\n",
    "        tc = 'w'\n",
    "    ax.arrow(12e6, y, 1.3e6, 0, color=cross_color[cross], lw=0, width=(dy[1]-dy[0])+0.9,\n",
    "             head_length=0, head_width=0, clip_on=False)\n",
    "    ax.text(12.65e6, y, f'{cross} ({denovo_counts_per_cross[cross]})', color=tc, ha='center', va='center', size=10)\n",
    "\n",
    "#Legend of Ty families\n",
    "legend_elms = [Line2D([0], [0], color='w', marker='^', ms=9, mfc=ty_colors[ty], mec=ty_colors[ty], label=ty_alias[ty]) for ty in ty_order[:-1]] + \\\n",
    "[Line2D([0], [0], color='w', marker='^', ms=9, mfc=mfc, mec='k', label=l) for (l, mfc) in zip(['parental', 'de novo'], ['w', 'k'])]\n",
    "\n",
    "ax.legend(handles=legend_elms, loc='lower center', bbox_to_anchor=[0.5, 1.01], ncols=8,\n",
    "          handlelength=1, handletextpad=0.5, columnspacing=1.2)\n",
    "    \n",
    "sns.despine()\n",
    "\n",
    "#for ext in ['svg', 'png']:\n",
    "#    plt.savefig(f'{fig_path}Fig4.{ext}', dpi=300)\n",
    "\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0ca93ce-08aa-4b50-a230-5e74ea7f1ba6",
   "metadata": {},
   "source": [
    "## Extract and align full-length Ty"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99f1b051-ea33-4ad3-b5a4-0ba1985d303c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_fasta_rea(rea, i, flag, pad=50):\n",
    "    \n",
    "    global fl_ty_seq\n",
    "    global genome\n",
    "    global s_subg\n",
    "    \n",
    "    ID, fam, query, start, end, strand = rea.loc[i, ['id', 'family', 'query', 'start', 'end', 'orient']]\n",
    "        \n",
    "    seq = genome[query].seq[start-1-pad:end+pad]\n",
    "    if strand == 'C':\n",
    "        seq = seq.reverse_complement()\n",
    "    name = f'{flag}.{s_subg}.{ID}.{fam}'\n",
    "    sr = SeqRecord.SeqRecord(seq, id=name, description='')\n",
    "\n",
    "    if fam not in fl_ty_seq:\n",
    "        fl_ty_seq[fam] = {}\n",
    "\n",
    "    if name not in fl_ty_seq[fam]:\n",
    "        fl_ty_seq[fam][name] = sr\n",
    "\n",
    "for flank, flank_alias in zip([0, 50, 10000], ['', '_flank50', '_flank10kb']):\n",
    "    \n",
    "    fl_ty_seq = {}\n",
    "    \n",
    "    # start with parental\n",
    "    \n",
    "    if flank != 10000:\n",
    "        for s_subg in Parents:\n",
    "\n",
    "            rea = REA.loc[(REA['s_subg']==s_subg) & (REA['fl'])]\n",
    "            genome = {seq.id:seq for seq in SeqIO.parse(f'/mnt/HDD1/paradoxus_nanopore/MA_parents/assemblies/{s_subg}.chromosomes.fasta', 'fasta')}\n",
    "\n",
    "            for i in rea.index:\n",
    "                extract_fasta_rea(rea, i, 'parent', pad=flank)\n",
    "\n",
    "\n",
    "    # continue with the de novo insertions            \n",
    "    for s_subg, df in og_denovo.groupby('s_subg'):\n",
    "\n",
    "        genome = {seq.id:seq for seq in SeqIO.parse(f'/mnt/HDD3/lrma/medaka/{s_subg}/consensus.fasta', 'fasta')}\n",
    "\n",
    "        for og in df['og'].unique():\n",
    "\n",
    "            rea = REA.loc[(REA['fl']) & (REA['s_subg']==s_subg) & (REA['og']==og)]\n",
    "\n",
    "            for i in rea.index:\n",
    "                if s_subg == 'J27.MSH-587-1' and rea.loc[i, 'id'] == 'i3':\n",
    "                    continue\n",
    "                else:\n",
    "                    extract_fasta_rea(rea, i, 'de_novo', pad=flank)\n",
    "\n",
    "    # export fasta sequences\n",
    "    for fam in fl_ty_seq:\n",
    "        if flank != 10000:\n",
    "            SeqIO.write(fl_ty_seq[fam].values(), f'/mnt/HDD3/lrma/sequences/{fam}.fl{flank_alias}.fasta', 'fasta')\n",
    "            ...\n",
    "        else:\n",
    "            for seq in fl_ty_seq[fam].values():\n",
    "                SeqIO.write(seq, f'/mnt/HDD3/lrma/sequences/{seq.id}.fl{flank_alias}.fasta', 'fasta')\n",
    "                ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9a1b351-9080-4fb5-ba50-08b84f93bc95",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract relevant representative sequences from Bleykasten-Grosshans et al. MBE 2021\n",
    "seq_mbe21 = {}\n",
    "coords_mbe21 = pd.read_csv('/mnt/HDD3/lrma/sequences/Ty1_mbe21.fl.csv', header=None)\n",
    "for i in coords_mbe21.index:\n",
    "    fam, contig, start, end, strand, description = coords_mbe21.loc[i]\n",
    "    seq = SeqIO.read(f'/mnt/HDD3/lrma/sequences/{contig}.fasta', 'fasta').seq[start-500:end+500]\n",
    "    if strand == '-':\n",
    "        seq = seq.reverse_complement()\n",
    "    sr = SeqRecord.SeqRecord(seq=seq, id=f'{fam}_mbe21', description='')\n",
    "    \n",
    "    seq_mbe21[sr.id] = sr\n",
    "    \n",
    "with open(f'/mnt/HDD3/lrma/sequences/Ty1_mbe21.fl_flank500.fasta', 'w') as handle:\n",
    "    SeqIO.write(seq_mbe21.values(), handle, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27715c53-373b-4e91-b8e7-685d02f6658d",
   "metadata": {},
   "source": [
    "## Target site duplications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bad5bc30-26b3-4b53-a2df-c4d1559e7e68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# parental TSDs - to be manually aligned and inspected visually\n",
    "seq_tsd_parents = []\n",
    "for p in Parents:\n",
    "\n",
    "        genome = {seq.id:seq for seq in SeqIO.parse(f'/mnt/HDD1/paradoxus_nanopore/MA_parents/assemblies/{p}.chromosomes.fasta', 'fasta')}\n",
    "\n",
    "        rea_sub = REA.loc[(REA['s_subg']==p) & (REA['fl']) & (REA['family']=='Ty1p')]\n",
    "        \n",
    "        for i in rea_sub.index:\n",
    "\n",
    "            rea = rea_sub.loc[i]\n",
    "            \n",
    "            query, start, end, orient = rea.loc[['query', 'start', 'end', 'orient']]\n",
    "            seq = genome[query].seq[start-11:start+11] + Seq.Seq('NNN') + genome[query].seq[end-11:end+11]\n",
    "            if orient == 'C':\n",
    "                seq = seq.reverse_complement()\n",
    "            \n",
    "            sr = SeqRecord.SeqRecord(seq=seq, id=f'{p}.{query}.{start:.0f}.{end:.0f}', description='')\n",
    "            seq_tsd_parents.append(sr)\n",
    "\n",
    "with open('/mnt/HDD3/lrma/sequences/seq_tsd_parents.fasta', 'w') as file:\n",
    "    SeqIO.write(seq_tsd_parents, file, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e757e40-1d43-4b02-a9dd-af43ab75c0b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "S_order = og_denovo.sort_values(by=['family','cross_order','s_subg']).set_index(['s_subg','og'], drop=False)\n",
    "S_order['order'] = range(S_order.shape[0])\n",
    "S_order['chrom'] = og_data.loc[S_order['og'], 'Query'].values\n",
    "S_order['label'] = S_order.apply(lambda x: f'{x[\"s_subg\"]} {x[\"chrom\"]}', axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4888f233-ac7d-43c2-a2a2-7b5a4f232b77",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_tsd = []\n",
    "for s_subg, df in og_denovo.groupby('s_subg'):\n",
    "\n",
    "        genome = {seq.id:seq for seq in SeqIO.parse(f'/mnt/HDD3/lrma/medaka/{s_subg}/consensus.fasta', 'fasta')}\n",
    "\n",
    "        for og in df['og'].unique():\n",
    "\n",
    "            rea = REA.loc[(REA['fl']) & (REA['s_subg']==s_subg) & (REA['og']==og)]\n",
    "            if og == 'og4121' and s_subg:\n",
    "                rea = rea.loc[rea['id']=='i4']\n",
    "            rea = rea.iloc[0]\n",
    "            \n",
    "            query, start, end, orient = rea.loc[['query', 'start', 'end', 'orient']]\n",
    "            seq = genome[query].seq[start-11:start+11] + Seq.Seq('NNN') + genome[query].seq[end-11:end+11]\n",
    "            if orient == 'C':\n",
    "                seq = seq.reverse_complement()\n",
    "            \n",
    "            sr = SeqRecord.SeqRecord(seq=seq, id=f'{s_subg},{og}', description='')\n",
    "            seq_tsd.append(sr)\n",
    "\n",
    "with open('/mnt/HDD3/lrma/sequences/seq_tsd.fasta', 'w') as file:\n",
    "    SeqIO.write(seq_tsd, file, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3d2728d-54fa-4f60-8336-336976eb7f3d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# import corrected sequences\n",
    "with open('/mnt/HDD3/lrma/sequences/seq_tsd.trim.fasta') as handle:\n",
    "    seq_tsd = {seq.id:seq for seq in SeqIO.parse(handle, 'fasta')}\n",
    "\n",
    "Seq_tsd = []\n",
    "for s_subg_og, seq in seq_tsd.items():\n",
    "    \n",
    "    #if s_subg_og != 'H33.UWOPS-91-202,og5200':\n",
    "    if s_subg_og in og_denovo.apply(lambda x: f'{x[\"s_subg\"]},{x[\"og\"]}', axis=1).values:\n",
    "        rearranged_seq = seq.seq.split('N')\n",
    "        left_side = rearranged_seq[0][-20:]\n",
    "        right_side = rearranged_seq[1][:20]\n",
    "        Seq_tsd.append(pd.Series(list(left_side) + list(right_side), name=s_subg_og,\n",
    "                       index=np.concatenate([np.arange(-20,0), np.arange(1,21)])))\n",
    "Seq_tsd = pd.concat(Seq_tsd, axis=1).T.reset_index()\n",
    "Seq_tsd[0] = '...'\n",
    "\n",
    "Seq_tsd = pd.melt(Seq_tsd, id_vars='index', var_name='pos', value_name='nt')\n",
    "for i in Seq_tsd.index:\n",
    "    s_subg, og = Seq_tsd.loc[i, 'index'].split(',')\n",
    "    Seq_tsd.loc[i, 's_order'] = S_order.loc[(s_subg, og), 'order']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31e957a9-6373-4dfe-9946-e544b9cb230e",
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_nt = {'A': '$\\mathtt{A}$', 'T':'$\\mathtt{T}$', 'C':'$\\mathtt{C}$', 'G':'$\\mathtt{G}$', '...':'$\\mathtt{...}$'}\n",
    "color_nt = {'A': 'limegreen', 'T':'red', 'C':'blue', 'G':'k', '...':'k'}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03b7e81a-3578-482b-a1fa-b9f34253321c",
   "metadata": {},
   "source": [
    "### Fig S7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08ab39d3-753d-4f90-a997-e7cac2b26d98",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=[8,5], gridspec_kw={'left':0.25, 'bottom':0.1, 'top':0.9, 'right':0.92})\n",
    "\n",
    "box_pad = 0.1\n",
    "\n",
    "for nt, df in Seq_tsd.groupby('nt'):\n",
    "    ax.scatter(df['pos'], df['s_order'], marker=marker_nt[nt], color=color_nt[nt], \n",
    "               linewidths=0.5, s=24, zorder=1)\n",
    "    \n",
    "rect = Rectangle((-10.5, -0.5), 21, 19, fc='0.85', zorder=0)\n",
    "ax.add_patch(rect)\n",
    "fa = FancyArrow(-10.5, 18.5, 21, 0, width=1, head_width=1.5, length_includes_head=True,\n",
    "               fc='white', ec='k', clip_on=False)\n",
    "ax.add_patch(fa)\n",
    "ax.text(0, 18.5, 'Full-length Ty1', ha='center', va='center')\n",
    "\n",
    "# mark identical TSD\n",
    "for s_subg_og, df in Seq_tsd.groupby('index'):\n",
    "    \n",
    "    s_subg, og = s_subg_og.split(',')\n",
    "    y = S_order.loc[(s_subg, og), 'order']\n",
    "    df = df.copy().set_index('pos')\n",
    "    \n",
    "    tsd_detected = False\n",
    "    \n",
    "    for tsd_len in [10,9,8,7,6,5,4,3]:\n",
    "        left_tsd_pos = list(range(-10-tsd_len, -10))\n",
    "        right_tsd_pos = list(range(11, 11+tsd_len))\n",
    "        left_tsd = ''.join(df.loc[left_tsd_pos, 'nt'].values)\n",
    "        right_tsd = ''.join(df.loc[right_tsd_pos, 'nt'].values)\n",
    "\n",
    "        if left_tsd == right_tsd:\n",
    "            tsd_detected = True\n",
    "            break\n",
    "    \n",
    "    if tsd_detected:\n",
    "        box_left = Rectangle((-10-tsd_len-0.5+box_pad, y-0.5+box_pad), tsd_len-2*box_pad, 1-2*box_pad,\n",
    "                             fc='#0f0f0f00', ec='k', lw=0.7)\n",
    "        box_right = Rectangle((10.5+box_pad, y-0.5+box_pad), tsd_len-2*box_pad, 1-2*box_pad,\n",
    "                             fc='#0f0f0f00', ec='k', lw=0.7)\n",
    "        ax.add_patch(box_left)\n",
    "        ax.add_patch(box_right)\n",
    "\n",
    "ax.set_xticks([-20, -15, -10, -5, 5, 10, 15, 20])\n",
    "ax.set_xticklabels(['-10', '-5', '0', '+5', '-5', '0', '+5', '+10'])\n",
    "ax.set_xlabel('Position (nt)')\n",
    "ax.set_xlim(-20.5, 20.5)\n",
    "ax.set_yticks(range(0,18))\n",
    "ax.set_yticklabels(S_order.set_index('order').loc[range(0,18), 'label'], size=8)\n",
    "ax.set_ylim(-0.5, 17.5)\n",
    "\n",
    "# add patches for crosses\n",
    "for cross, dy in zip(['CC1', 'CC2', 'CC3', 'BB1', 'BC2', 'BA1', 'BA2', 'BSc1'],\n",
    "                     [(0,3), (4,5), (6,8), (9,9), (10,12), (13,13), (14,16), (17,17)]):\n",
    "    \n",
    "    y = np.mean(dy)\n",
    "    if cross == 'BSc1':\n",
    "        tc = 'k'\n",
    "    else:\n",
    "        tc = 'w'\n",
    "    ax.arrow(21, y, 3, 0, color=cross_color[cross], lw=0, width=(dy[1]-dy[0])+0.9,\n",
    "             head_length=0, head_width=0, clip_on=False)\n",
    "    ax.text(22.5, y, cross, color=tc, ha='center', va='center', size=10)\n",
    "\n",
    "sns.despine()\n",
    "    \n",
    "#for ext in ['svg', 'png']:\n",
    "#    plt.savefig(f'{fig_path}Supp_FigS7.{ext}', dpi=300)\n",
    "\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76de3a87-2914-4e2c-906c-1c0e4f7ecaf8",
   "metadata": {},
   "source": [
    "## Phylogenetic analysis of full-length Tys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b31e139b-c799-4163-8ad3-b6a4e4a8f7d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset bootstrap trees for manual examination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8957d883-52cd-4823-bb08-15e8cadfdd40",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/mnt/HDD3/lrma/muscle/Ty1.fl_flank50.aln.fil.fasta.raxml.bootstraps') as handle:\n",
    "    bootstrap_trees = handle.read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bb87b69-8142-4f23-82e2-f89876237c04",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export \n",
    "random.seed(42)\n",
    "for i in random.sample(range(len(bootstrap_trees)), 10):\n",
    "    with open(f'/mnt/HDD3/lrma/muscle/Ty1.fl_flank50.aln.fil.fasta.raxml.bootstraps_{i}.nwk', 'w') as handle:\n",
    "        handle.write(bootstrap_trees[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac75cdd1-f269-49e0-8e96-fbef87ada81d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert fasta alignment to nexus for SplitsTree\n",
    "fasta = SeqIO.convert('/mnt/HDD3/lrma/muscle/Ty1.fl_flank50.aln.fil.fasta', 'fasta',\n",
    "                      '/mnt/HDD3/lrma/muscle/Ty1.fl_flank50.aln.fil.nxs', 'nexus', 'DNA')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
